{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/project/cs-myers/MathewF/projects/Laura-SB-Analysis/NetCIS/notebooks')"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn.objects as so\n",
    "from seaborn import axes_style\n",
    "import networkx as nx\n",
    "from scipy.stats import chisquare, binomtest, ranksums, mannwhitneyu, skewtest, kurtosistest\n",
    "\n",
    "from IPython.display import display\n",
    "Path.cwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_subgraphs(graph_dir, graph_type):\n",
    "    subgraph_dict = {}\n",
    "    for graph in (graph_dir / graph_type).iterdir():\n",
    "        chrom = graph.name.split(\".\")[0]\n",
    "        G = nx.read_graphml(graph)\n",
    "        subgraphs_by_nodes = sorted(nx.connected_components(G), key=len, reverse=True)\n",
    "        subgraph_dict[chrom] = [ G.subgraph(x) for x in subgraphs_by_nodes ]\n",
    "    return subgraph_dict\n",
    "\n",
    "def graph_properties(G, verbose=0):\n",
    "    nodes = G.number_of_nodes()\n",
    "    edges = G.number_of_edges()\n",
    "    num_inserts = sum([ G.nodes[node]['counts'] for node in G.nodes ])\n",
    "    subgraphs_by_nodes = sorted(nx.connected_components(G), key=len, reverse=True)\n",
    "    num_subgraphs = len(subgraphs_by_nodes)\n",
    "    if verbose:\n",
    "        print(f\"number of nodes: {nodes}\")\n",
    "        print(f\"number of edges: {edges}\")\n",
    "        print(f\"number of insertions: {num_inserts}\")\n",
    "        print(f\"number of subgraphs (pCIS) {num_subgraphs}\")\n",
    "    return {\"nodes\": nodes, \"edges\": edges, \"num_inserts\": num_inserts, \"num_subgraphs\": num_subgraphs}\n",
    "\n",
    "def subgraph_properties(G, verbose=0):\n",
    "    nodes = G.number_of_nodes()\n",
    "    edges = G.number_of_edges()\n",
    "    num_inserts = sum([ G.nodes[node]['counts'] for node in G.nodes ])\n",
    "    tmp_pos = sorted([ G.nodes[node][\"position\"] for node in G.nodes ])\n",
    "    min_pos = min(tmp_pos)\n",
    "    max_pos = max(tmp_pos)\n",
    "    range_pos = max_pos - min_pos\n",
    "    if verbose:\n",
    "        print(f\"number of nodes: {nodes}\")\n",
    "        print(f\"number of edges: {edges}\")\n",
    "        print(f\"number of insertions: {num_inserts}\")\n",
    "        print(f\"min position: {min_pos}\")\n",
    "        print(f\"max position: {max_pos}\")\n",
    "        print(f\"range: {range_pos}\")\n",
    "    # TODO: FIXME: total nodes and range are not inclusive. see chr7\n",
    "    # 7/10/23 - what does this mean...?\n",
    "    return {\"nodes\": nodes, \"edges\": edges, \"num_inserts\": num_inserts, \"min_pos\": min_pos, \"max_pos\": max_pos, \"range\": range_pos}\n",
    "\n",
    "def subgraph_TA_sites(G, bed, ta_error, verbose=0):\n",
    "    num_insert_sites = G.number_of_nodes()\n",
    "    \n",
    "    tmp_pos = sorted([ G.nodes[node][\"position\"] for node in G.nodes ])\n",
    "    ta_sites = bed[(bed[1] > min(tmp_pos)) & (bed[2] < max(tmp_pos))]\n",
    "    num_ta_sites = len(ta_sites)\n",
    "    \n",
    "    arr1 = np.array(tmp_pos).reshape(-1, 1)\n",
    "    arr2 = ta_sites[1].to_numpy().reshape(-1, 1)\n",
    "    arr3 = ta_sites[1].to_numpy().reshape(-1, 1)\n",
    "    ta_inserts = (arr1 >= (arr2.T - ta_error)) & (arr1 <= (arr3.T + ta_error))\n",
    "    num_ta_insert_sites = ta_inserts.any(axis=1).sum()\n",
    "    \n",
    "    if verbose:\n",
    "        print(f\"number of insertions in subgraph: {num_insert_sites}\")\n",
    "        print(f\"number of TA sites in subgraph: {num_ta_sites}\")\n",
    "        print(f\"number of insertions within a TA site (+/- {ta_error} bp): {num_ta_insert_sites}\")\n",
    "        \n",
    "    return {\"num_insert_sites\": num_insert_sites, \"num_ta_sites\": num_ta_sites, \"num_ta_insert_sites\": num_ta_insert_sites}\n",
    "\n",
    "def get_subgraph_stats(subgraph_chroms, graph_type, bed_files, ta_error):\n",
    "    subgraph_df_list = []\n",
    "    for chrom, subgraphs in subgraph_chroms.items():\n",
    "        for i, subgraph in enumerate(subgraphs):\n",
    "            sg_meta = {\"type\": graph_type, \"chrom\": chrom, \"subgraph\": i}\n",
    "            sg_prop = subgraph_properties(subgraph)\n",
    "            sg_ta = subgraph_TA_sites(subgraph, bed_files[chrom], ta_error)\n",
    "            sg_df = pd.DataFrame((sg_meta | sg_prop | sg_ta))\n",
    "            subgraph_df_list.append(sg_df)\n",
    "    return pd.concat(subgraph_df_list, ignore_index=True)  \n",
    "\n",
    "# def get_graph(graph_dir, graph_type, bed_files, ta_error):\n",
    "#     case_subgraphs = {}\n",
    "#     case_df_list = []\n",
    "#\n",
    "#     for graph in (graph_dir / graph_type).iterdir():\n",
    "#         chrom = graph.name.split(\".\")[0]\n",
    "#         G = nx.read_graphml(graph)\n",
    "#         subgraphs_by_nodes = sorted(nx.connected_components(G), key=len, reverse=True)\n",
    "#         subgraph_list = []\n",
    "#         for i, x in enumerate(subgraphs_by_nodes):\n",
    "#             sg_meta = {\"type\": graph_type, \"chrom\": chrom, \"subgraph\": i}\n",
    "#             subgraph = G.subgraph(x)\n",
    "#             subgraph_list.append(subgraph)\n",
    "#             sg_prop = subgraph_properties(subgraph)\n",
    "#             sg_ta = subgraph_TA_sites(subgraph, bed_files[chrom], ta_error)\n",
    "#             sg_df = pd.DataFrame((sg_meta | sg_prop | sg_ta), index=[i])\n",
    "#             case_df_list.append(sg_df)\n",
    "#         case_subgraphs[chrom] = subgraph_list\n",
    "#     case_df = pd.concat(case_df_list, ignore_index=True)\n",
    "#     return case_df, case_subgraphs\n",
    "    \n",
    "def pcis_overlaps(target_df, reference_df):\n",
    "    # for all pCISs in target_df, find any overlap with reference_df (could be more than one or none)\n",
    "    \n",
    "    # start with each target subgraph, look through all reference subgraphs and record any overlap in the positions\n",
    "    # key: index of target subgraph, value: list of reference subgraph indexes that overlap within the range of target subgraph\n",
    "    overlap_dict = {}\n",
    "    for tar_sg in target_df.itertuples():\n",
    "        overlap_list = []\n",
    "        for ref_sg in reference_df.itertuples():\n",
    "            if (tar_sg.min_pos <= ref_sg.max_pos) and (tar_sg.max_pos >= ref_sg.min_pos):\n",
    "                overlap_list.append(ref_sg.subgraph)\n",
    "        overlap_dict[tar_sg.subgraph] = overlap_list\n",
    "    # TODO: would it be easier/more efficient to compute this by whole dataframe?\n",
    "    # this is the fully coded out logic, but now can I make it better and improve my pandas skills?\n",
    "    # I can use chrom_overlaps as the ground truth as well.\n",
    "    \n",
    "    return overlap_dict\n",
    "\n",
    "def compare_pcis(target_overlaps, target_subgraphs, reference_subgraphs):\n",
    "    # do case-control comparison of TAs between overlapping subgraphs (pCIS):\n",
    "    # - Match TA to TA site per subgraph\n",
    "    # - calculate log fold changes\n",
    "    # - use binomtest for significance of each TA\n",
    "    # - Then use overall statistic for independent sample test between target and all references\n",
    "    # which can be used for the final determination if the pCIS is now a CIS\n",
    "    TA_df_list = []\n",
    "    overall_df_list = []\n",
    "\n",
    "    for tar_ind, ref_inds in target_overlaps.items():\n",
    "        tar_G = target_subgraphs[tar_ind]\n",
    "        tar_pos = [ tar_G.nodes[node]['position'] for node in tar_G.nodes ]\n",
    "        tmp_tar = pd.DataFrame([ {\"target\": tar_G.nodes[node]['counts']} for node in tar_G.nodes ], index=tar_pos)        \n",
    "        \n",
    "        if len(ref_inds) == 0:\n",
    "            tmp = tmp_tar\n",
    "            tmp[\"reference\"] = 0\n",
    "            tmp[\"reference_index\"] = None\n",
    "        else:\n",
    "            tmp_ref_list = []\n",
    "            for ref_ind in ref_inds:\n",
    "                ref_G = reference_subgraphs[ref_ind]\n",
    "                ref_pos = [ ref_G.nodes[node]['position'] for node in ref_G.nodes ]\n",
    "                ref_cols = [ {\"reference\": ref_G.nodes[node]['counts'], \"reference_index\": ref_ind} for node in ref_G.nodes ]\n",
    "                tmp_ref_list.append(pd.DataFrame(ref_cols, index=ref_pos))\n",
    "                \n",
    "            # get union of all insertion sites and make it into a df\n",
    "            tmp_ref = pd.concat(tmp_ref_list)\n",
    "            tmp = tmp_tar.join(tmp_ref, how=\"outer\")\n",
    "    \n",
    "        tmp = tmp.reset_index(drop=False).rename(columns={\"index\": \"pos\"})\n",
    "        tmp[\"target_index\"] = tar_ind\n",
    "        tmp[\"target\"] = tmp[\"target\"].fillna(0).astype(int)\n",
    "        tmp[\"reference\"] = tmp[\"reference\"].fillna(0).astype(int)\n",
    "        \n",
    "        # get stats per TA site (only count is used)\n",
    "        tmp[\"target_binom_pval\"] = tmp.apply(lambda x: binomtest(int(x[\"target\"]), int(x[\"target\"] + x[\"reference\"])).pvalue, axis=1)\n",
    "        tmp[\"target_binom_sig\"] = tmp[\"target_binom_pval\"] < 0.05\n",
    "        tmp[\"LFC\"] = tmp.apply(lambda x: np.log2((x[\"target\"]+1) / (x[\"reference\"]+1)), axis=1)\n",
    "        # used pseudo count of 1 for log fold change, and so I wanted to show the difference in binomial test and significance with this\n",
    "        tmp[\"p_target_binom_pval\"] = tmp.apply(lambda x: binomtest(x[\"target\"] + 1, x[\"target\"] + x[\"reference\"] + 2).pvalue, axis=1)\n",
    "        tmp[\"p_target_binom_sig\"] = tmp[\"p_target_binom_pval\"] < 0.05\n",
    "        \n",
    "        # overall test stat for independence. use genomic positions. Total samples are each position times counts.\n",
    "        # ex.) pos: 1001 and count: 3 is [1001, 1001, 1001]\n",
    "        target_overall = []\n",
    "        reference_overall = []\n",
    "        for row in tmp.itertuples():\n",
    "            for pos_tmp in [row.pos] * row.target:\n",
    "                target_overall.append(pos_tmp)\n",
    "            for pos_tmp in [row.pos] * row.reference:\n",
    "                reference_overall.append(pos_tmp)\n",
    "\n",
    "        mwu = mannwhitneyu(target_overall, reference_overall).pvalue if len(reference_overall) != 0 else np.nan\n",
    "        rs = ranksums(target_overall, reference_overall).pvalue if len(reference_overall) != 0 else np.nan\n",
    "        # case_skewtest = skewtest(target_overall).pvalue if len(target_overall) >= 8 else np.nan\n",
    "        # case_kurtosistest = kurtosistest(target_overall).pvalue if len(target_overall) >= 20 else np.nan\n",
    "        # control_skewtest = skewtest(reference_overall).pvalue if len(reference_overall) >= 8 else np.nan\n",
    "        # control_kurtosistest = kurtosistest(reference_overall).pvalue if len(reference_overall) >= 20 else np.nan\n",
    "        total_TA = len(tmp)\n",
    "        TA_sig = tmp[\"target_binom_sig\"].sum()\n",
    "                            \n",
    "        # Other stats: Kurtosis, skewness, etc. # \"stat type\": [\"statistic\", \"pvalue\"],\n",
    "        tmp2 = pd.DataFrame({\n",
    "            \"target_index\": [tar_ind],\n",
    "            # \"reference_index\": [tmp[\"reference_index\"].values[0]],\n",
    "            \"target_pos_min\": [min(tar_pos)],\n",
    "            \"target_pos_max\": [max(tar_pos)],\n",
    "            # \"reference_pos_min\": [min(ref_pos)] if len(ref_inds) != 0 else [None],\n",
    "            # \"reference_pos_max\": [max(ref_pos)] if len(ref_inds) != 0 else [None],\n",
    "            \"mannwhitneyu\": [mwu],\n",
    "            \"ranksums\": [rs], \n",
    "            #  \"case-skewtest\": [case_skewtest],\n",
    "            #  \"case-kurtosistest\": [case_kurtosistest],\n",
    "            #  \"control-skewtest\": [control_skewtest],\n",
    "            #  \"control-kurtosistest\": [control_kurtosistest],\n",
    "            \"total_TA_sites\": [total_TA],\n",
    "            \"sig_TA_sites\": [TA_sig],\n",
    "            })\n",
    "\n",
    "        TA_df_list.append(tmp)\n",
    "        overall_df_list.append(tmp2)\n",
    "\n",
    "    TA_df = pd.concat(TA_df_list, ignore_index=True)\n",
    "    overall_df = pd.concat(overall_df_list, ignore_index=True)\n",
    "    overall_df[\"sig_ratio\"] = overall_df[\"sig_TA_sites\"] / overall_df[\"total_TA_sites\"]\n",
    "    return TA_df, overall_df\n",
    "\n",
    "def pcis_to_cis(overall_df, threshold):\n",
    "    # find pcis with significant pvalue that is less than the given threshold\n",
    "    sig_df = overall_df[ (overall_df[\"mannwhitneyu\"] < threshold) & (overall_df[\"ranksums\"] < threshold) ]\n",
    "    # test stat below threshold OR ratio that is not 0 and there are more than 1 sig tA\n",
    "    nan_df =  overall_df[ pd.isna(overall_df[\"mannwhitneyu\"]) & pd.isna(overall_df[\"ranksums\"]) ]\n",
    "    nan_sig_df = nan_df[ (nan_df[\"sig_ratio\"] != 0) & (nan_df[\"sig_TA_sites\"] > 1) ]\n",
    "    all_sig_df = pd.concat([sig_df, nan_sig_df])\n",
    "    return all_sig_df\n",
    "\n",
    "def cis_annotate(target_sig_df, annotated_df):\n",
    "    # get genes markers to each CIS\n",
    "    gene_list = []\n",
    "    for row in target_sig_df.itertuples():\n",
    "        sub_gene_list = []\n",
    "        for gene in annotated_df.itertuples():\n",
    "            try:\n",
    "                sub3 = row.reference_pos_min <= gene._5\n",
    "                sub4 = row.reference_pos_max >= gene._4\n",
    "            except:\n",
    "                sub3 = None\n",
    "                sub4 = None\n",
    "            # TODO: just use the whole row from the annotation\n",
    "            if (row.target_pos_min <= gene._5) and (row.target_pos_max >= gene._4):\n",
    "                sub_gene_list.append(pd.DataFrame({\n",
    "                    \"type\": [\"target\"],\n",
    "                    \"type_index\": [int(row.target_index)],\n",
    "                    \"marker_pos_start\": [gene._4],\n",
    "                    \"marker_pos_end\": [gene._5],\n",
    "                    # \"marker_strand\": [gene.strand],\n",
    "                    \"marker_symbol\": [gene._7],\n",
    "                    \"marker_name\": [gene._8],\n",
    "                    \"marker_type\": [gene._9],\n",
    "                    \"marker_feature_type\": [gene._10],\n",
    "                    \"marker_annot_index\": [gene.Index],\n",
    "                    }))\n",
    "            elif sub3 and sub4:\n",
    "                sub_gene_list.append(pd.DataFrame({\n",
    "                    \"type\": [\"reference\"],\n",
    "                    \"type_index\": [int(row.reference_index)],\n",
    "                    \"marker_pos_start\": [gene._4],\n",
    "                    \"marker_pos_end\": [gene._5],\n",
    "                    # \"marker_strand\": [gene.strand],\n",
    "                    \"marker_symbol\": [gene._7],\n",
    "                    \"marker_name\": [gene._8],\n",
    "                    \"marker_type\": [gene._9],\n",
    "                    \"marker_feature_type\": [gene._10],\n",
    "                    \"marker_annot_index\": [gene.Index],\n",
    "                    }))\n",
    "                \n",
    "        # it is possible that there were no annotations found\n",
    "        if len(sub_gene_list) == 0:\n",
    "            sub_gene_list.append(pd.DataFrame({\n",
    "                \"type\": [\"target\"],\n",
    "                \"type_index\": [int(row.target_index)],\n",
    "                \"marker_pos_start\": [None],\n",
    "                \"marker_pos_end\": [None],\n",
    "                # \"marker_strand\": [None],\n",
    "                \"marker_symbol\": [None],\n",
    "                \"marker_name\": [None],\n",
    "                \"marker_type\": [None],\n",
    "                \"marker_feature_type\": [None],\n",
    "                \"marker_annot_index\": [None],\n",
    "                }))\n",
    "            \n",
    "        gene_list.extend(sub_gene_list)\n",
    "    return pd.concat(gene_list, ignore_index=True)\n",
    "                \n",
    "def volcano_plot(data, lfc, pval, threshold=0.05):\n",
    "    thres = np.log10(threshold) * -1\n",
    "    data[pval] = np.log10(data[pval]) * -1\n",
    "    g = (\n",
    "        so.Plot(data, x=lfc, y=pval, pointsize=pval)\n",
    "        # .add(so.Dots(), so.Jitter(1))\n",
    "        .add(so.Dots(color=\"grey\"), so.Jitter(1), data=data.query(f\"{pval} < {thres}\"))\n",
    "        .add(so.Dots(color=\"blue\"), so.Jitter(1), data=data.query(f\"{pval} >= {thres}\"))\n",
    "        .scale(y=\"log\")\n",
    "    )\n",
    "    return g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_dir = Path(\"/project/cs-myers/MathewF/projects/Laura-SB-Analysis/output/2020_SB-graphs/\")\n",
    "ta_dir = Path(\"/project/cs-myers/MathewF/software/bowtie2-2.4.5/indexes/GRCm39_TAs/\")\n",
    "gene_annot = Path(\"/project/cs-myers/MathewF/projects/Laura-SB-Analysis/NetCIS/toy-data/MRK_List2.rpt\")\n",
    "output_prefix = \"/project/cs-myers/MathewF/projects/Laura-SB-Analysis/output/2020_SB\"\n",
    "output = Path(output_prefix + \"-analysis\")\n",
    "output.mkdir(exist_ok=True)\n",
    "ta_error = 5\n",
    "pval_threshold = 0.05\n",
    "verbose = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "annot_df = pd.read_csv(gene_annot, sep=\"\\t\")\n",
    "annot_df = annot_df[pd.notna(annot_df[\"genome coordinate start\"])].drop(\"Status\", axis=1)\n",
    "annot_df[\"chrom\"] = annot_df[\"Chr\"].apply(lambda x: f\"chr{x}\")\n",
    "annot_df = annot_df.sort_values([\"chrom\"]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI Accession ID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>cM Position</th>\n",
       "      <th>genome coordinate start</th>\n",
       "      <th>genome coordinate end</th>\n",
       "      <th>strand</th>\n",
       "      <th>Marker Symbol</th>\n",
       "      <th>Marker Name</th>\n",
       "      <th>Marker Type</th>\n",
       "      <th>Feature Type</th>\n",
       "      <th>Marker Synonyms (pipe-separated)</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74139</th>\n",
       "      <td>MGI:1926137</td>\n",
       "      <td>11</td>\n",
       "      <td>2.19</td>\n",
       "      <td>3081850.0</td>\n",
       "      <td>3143463.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Sfi1</td>\n",
       "      <td>Sfi1 homolog, spindle assembly associated (yeast)</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>2310047I15Rik</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      MGI Accession ID Chr cM Position  genome coordinate start  \\\n",
       "74139      MGI:1926137  11        2.19                3081850.0   \n",
       "\n",
       "       genome coordinate end strand Marker Symbol  \\\n",
       "74139              3143463.0      -          Sfi1   \n",
       "\n",
       "                                             Marker Name Marker Type  \\\n",
       "74139  Sfi1 homolog, spindle assembly associated (yeast)        Gene   \n",
       "\n",
       "              Feature Type Marker Synonyms (pipe-separated)  chrom  \n",
       "74139  protein coding gene                    2310047I15Rik  chr11  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# annot_df[annot_df[\"Marker Symbol\"] == \"Sfi1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_files = { file.name.split(\".\")[0]: pd.read_csv(file, sep=\"\\t\", header=None) for file in ta_dir.iterdir() }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "UFuncTypeError",
     "evalue": "ufunc 'subtract' did not contain a loop with signature matching types (dtype('int64'), dtype('<U8')) -> None",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUFuncTypeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m case_subgraph_dict \u001b[39m=\u001b[39m get_subgraphs(graph_dir, \u001b[39m\"\u001b[39m\u001b[39mcase\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m----> 2\u001b[0m case_df \u001b[39m=\u001b[39m get_subgraph_stats(case_subgraph_dict, \u001b[39m\"\u001b[39;49m\u001b[39mcase\u001b[39;49m\u001b[39m\"\u001b[39;49m, bed_files, \u001b[39m\"\u001b[39;49m\u001b[39mta_error\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n\u001b[1;32m      4\u001b[0m control_subgraph_dict \u001b[39m=\u001b[39m get_subgraphs(graph_dir, \u001b[39m\"\u001b[39m\u001b[39mcontrol\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m      5\u001b[0m control_df \u001b[39m=\u001b[39m get_subgraph_stats(control_subgraph_dict, \u001b[39m\"\u001b[39m\u001b[39mcontrol\u001b[39m\u001b[39m\"\u001b[39m, bed_files, \u001b[39m\"\u001b[39m\u001b[39mta_error\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "Cell \u001b[0;32mIn[2], line 68\u001b[0m, in \u001b[0;36mget_subgraph_stats\u001b[0;34m(subgraph_chroms, graph_type, bed_files, ta_error)\u001b[0m\n\u001b[1;32m     66\u001b[0m sg_meta \u001b[39m=\u001b[39m {\u001b[39m\"\u001b[39m\u001b[39mtype\u001b[39m\u001b[39m\"\u001b[39m: graph_type, \u001b[39m\"\u001b[39m\u001b[39mchrom\u001b[39m\u001b[39m\"\u001b[39m: chrom, \u001b[39m\"\u001b[39m\u001b[39msubgraph\u001b[39m\u001b[39m\"\u001b[39m: i}\n\u001b[1;32m     67\u001b[0m sg_prop \u001b[39m=\u001b[39m subgraph_properties(subgraph)\n\u001b[0;32m---> 68\u001b[0m sg_ta \u001b[39m=\u001b[39m subgraph_TA_sites(subgraph, bed_files[chrom], ta_error)\n\u001b[1;32m     69\u001b[0m sg_df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame((sg_meta \u001b[39m|\u001b[39m sg_prop \u001b[39m|\u001b[39m sg_ta))\n\u001b[1;32m     70\u001b[0m subgraph_df_list\u001b[39m.\u001b[39mappend(sg_df)\n",
      "Cell \u001b[0;32mIn[2], line 52\u001b[0m, in \u001b[0;36msubgraph_TA_sites\u001b[0;34m(G, bed, ta_error, verbose)\u001b[0m\n\u001b[1;32m     50\u001b[0m arr2 \u001b[39m=\u001b[39m ta_sites[\u001b[39m1\u001b[39m]\u001b[39m.\u001b[39mto_numpy()\u001b[39m.\u001b[39mreshape(\u001b[39m-\u001b[39m\u001b[39m1\u001b[39m, \u001b[39m1\u001b[39m)\n\u001b[1;32m     51\u001b[0m arr3 \u001b[39m=\u001b[39m ta_sites[\u001b[39m1\u001b[39m]\u001b[39m.\u001b[39mto_numpy()\u001b[39m.\u001b[39mreshape(\u001b[39m-\u001b[39m\u001b[39m1\u001b[39m, \u001b[39m1\u001b[39m)\n\u001b[0;32m---> 52\u001b[0m ta_inserts \u001b[39m=\u001b[39m (arr1 \u001b[39m>\u001b[39m\u001b[39m=\u001b[39m (arr2\u001b[39m.\u001b[39;49mT \u001b[39m-\u001b[39;49m ta_error)) \u001b[39m&\u001b[39m (arr1 \u001b[39m<\u001b[39m\u001b[39m=\u001b[39m (arr3\u001b[39m.\u001b[39mT \u001b[39m+\u001b[39m ta_error))\n\u001b[1;32m     53\u001b[0m num_ta_insert_sites \u001b[39m=\u001b[39m ta_inserts\u001b[39m.\u001b[39many(axis\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m)\u001b[39m.\u001b[39msum()\n\u001b[1;32m     55\u001b[0m \u001b[39mif\u001b[39;00m verbose:\n",
      "\u001b[0;31mUFuncTypeError\u001b[0m: ufunc 'subtract' did not contain a loop with signature matching types (dtype('int64'), dtype('<U8')) -> None"
     ]
    }
   ],
   "source": [
    "case_subgraph_dict = get_subgraphs(graph_dir, \"case\")\n",
    "case_df = get_subgraph_stats(case_subgraph_dict, \"case\", bed_files, \"ta_error\")\n",
    "\n",
    "control_subgraph_dict = get_subgraphs(graph_dir, \"control\")\n",
    "control_df = get_subgraph_stats(control_subgraph_dict, \"control\", bed_files, \"ta_error\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chroms = case_df[\"chrom\"].sort_values().unique()\n",
    "all_sig_annot_list = []\n",
    "all_sg_annot_list = []\n",
    "# TODO: parallelize this?\n",
    "for chrom in chroms:\n",
    "    # if chrom != \"chr6\":\n",
    "    #     continue\n",
    "    print(chrom)\n",
    "    \n",
    "    # get chromosome subsets for case and control\n",
    "    case_chrom_df = case_df[case_df[\"chrom\"] == chrom]\n",
    "    control_chrom_df = control_df[control_df[\"chrom\"] == chrom]\n",
    "    case_chrom_subgraphs = case_subgraph_dict[chrom]\n",
    "    control_chrom_subgraphs = control_subgraph_dict[chrom]\n",
    "    annot_chrom_df = annot_df[annot_df[\"chrom\"] == chrom]\n",
    "    \n",
    "    # cases as the target\n",
    "    case_overlaps = pcis_overlaps(case_chrom_df, control_chrom_df)\n",
    "    case_TA_df, case_overall_df = compare_pcis(case_overlaps, case_chrom_subgraphs, control_chrom_subgraphs)\n",
    "    case_sig_df = pcis_to_cis(case_overall_df, pval_threshold)\n",
    "    if len(case_sig_df) != 0:\n",
    "        case_sig_annot = cis_annotate(case_sig_df, annot_chrom_df)\n",
    "        case_sig_annot[\"class\"] = \"case\"\n",
    "        # keep only genes annotated to each significant subgraph\n",
    "        case_sg_annot = case_sig_df.join(case_sig_annot[case_sig_annot[\"marker_type\"] == \"Gene\"].groupby(\"type_index\").agg(list)[\"marker_symbol\"])\n",
    "        case_sg_annot[\"class\"] = \"case\"\n",
    "    else:\n",
    "        case_sig_annot = None\n",
    "    \n",
    "    # controls as the target\n",
    "    control_overlaps = pcis_overlaps(control_chrom_df, case_chrom_df)\n",
    "    control_TA_df, control_overall_df = compare_pcis(control_overlaps, control_chrom_subgraphs, case_chrom_subgraphs)\n",
    "    control_sig_df = pcis_to_cis(control_overall_df, pval_threshold)\n",
    "    if len(control_sig_df) != 0:\n",
    "        control_sig_annot = cis_annotate(control_sig_df, annot_chrom_df)\n",
    "        control_sig_annot[\"class\"] = \"control\"\n",
    "        control_sg_annot = control_sig_df.join(control_sig_annot[control_sig_annot[\"marker_type\"] == \"Gene\"].groupby(\"type_index\").agg(list)[\"marker_symbol\"])\n",
    "        control_sg_annot[\"class\"] = \"control\"\n",
    "    else:\n",
    "        control_annot = None\n",
    "    \n",
    "    if case_sig_annot is not None and control_sig_annot is not None:\n",
    "        both_sig_annot = pd.concat([case_sig_annot, control_sig_annot], ignore_index=True)\n",
    "        both_sg_annot = pd.concat([case_sg_annot, control_sg_annot], ignore_index=True)\n",
    "    elif case_sig_annot is not None:\n",
    "        both_sig_annot = case_sig_annot\n",
    "        both_sg_annot = case_sg_annot\n",
    "    elif control_sig_annot is not None:\n",
    "        both_sig_annot = control_sig_annot\n",
    "        both_sg_annot = control_sg_annot\n",
    "    else:\n",
    "        continue\n",
    "    \n",
    "    if case_sig_annot is not None or control_sig_annot is not None:\n",
    "        both_sig_annot[\"chrom\"] = chrom\n",
    "        all_sig_annot_list.append(both_sig_annot)\n",
    "        both_sg_annot[\"chrom\"] = chrom\n",
    "        all_sg_annot_list.append(both_sg_annot)\n",
    "        if verbose > 0:\n",
    "            print(len(both_sig_annot[\"marker_symbol\"].unique()))\n",
    "\n",
    "    # TODO: are there too many repeated genes? Does this make sense that they would be repeated?\n",
    "    # it appears that sometimes there are multiple CIS in a gene, because the CIS range is quite small\n",
    "    print(f\"\"\"\\tsig. genomic features: {both_sig_annot[\"marker_symbol\"].unique().shape[0]}/{annot_chrom_df[\"Marker Symbol\"].unique().shape[0]}\"\"\")\n",
    "\n",
    "\n",
    "all_sig_annot_df = pd.concat(all_sig_annot_list, ignore_index=True)\n",
    "all_sg_annot_df = pd.concat(all_sg_annot_list, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_index</th>\n",
       "      <th>target_pos_min</th>\n",
       "      <th>target_pos_max</th>\n",
       "      <th>mannwhitneyu</th>\n",
       "      <th>ranksums</th>\n",
       "      <th>total_TA_sites</th>\n",
       "      <th>sig_TA_sites</th>\n",
       "      <th>sig_ratio</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>class</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>22891694</td>\n",
       "      <td>22978601</td>\n",
       "      <td>2.319250e-02</td>\n",
       "      <td>2.471693e-02</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>[Gm7759]</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>29559897</td>\n",
       "      <td>29629191</td>\n",
       "      <td>1.860003e-02</td>\n",
       "      <td>1.756369e-02</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[slck]</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>26556902</td>\n",
       "      <td>26697890</td>\n",
       "      <td>4.746838e-15</td>\n",
       "      <td>1.785891e-14</td>\n",
       "      <td>17</td>\n",
       "      <td>4</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>[slck, Gm31136, Gm24064]</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>102741240</td>\n",
       "      <td>102765559</td>\n",
       "      <td>2.742386e-02</td>\n",
       "      <td>2.830225e-02</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>59</td>\n",
       "      <td>6047345</td>\n",
       "      <td>6163765</td>\n",
       "      <td>4.096985e-08</td>\n",
       "      <td>6.809029e-07</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>NaN</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>37</td>\n",
       "      <td>99977058</td>\n",
       "      <td>100032516</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>[Stol, Tex11]</td>\n",
       "      <td>case</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>45</td>\n",
       "      <td>57714231</td>\n",
       "      <td>57729530</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>case</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>66</td>\n",
       "      <td>47592563</td>\n",
       "      <td>47599471</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[Aifm1, Fnld]</td>\n",
       "      <td>case</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>17</td>\n",
       "      <td>73543988</td>\n",
       "      <td>73543996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[Stol]</td>\n",
       "      <td>control</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>17</td>\n",
       "      <td>73543988</td>\n",
       "      <td>73543996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[Stol]</td>\n",
       "      <td>control</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>537 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     target_index  target_pos_min  target_pos_max  mannwhitneyu      ranksums  \\\n",
       "0               4        22891694        22978601  2.319250e-02  2.471693e-02   \n",
       "1               5        29559897        29629191  1.860003e-02  1.756369e-02   \n",
       "2               6        26556902        26697890  4.746838e-15  1.785891e-14   \n",
       "3              10       102741240       102765559  2.742386e-02  2.830225e-02   \n",
       "4              59         6047345         6163765  4.096985e-08  6.809029e-07   \n",
       "..            ...             ...             ...           ...           ...   \n",
       "532            37        99977058       100032516           NaN           NaN   \n",
       "533            45        57714231        57729530           NaN           NaN   \n",
       "534            66        47592563        47599471           NaN           NaN   \n",
       "535            17        73543988        73543996           NaN           NaN   \n",
       "536            17        73543988        73543996           NaN           NaN   \n",
       "\n",
       "     total_TA_sites  sig_TA_sites  sig_ratio             marker_symbol  \\\n",
       "0                14             8   0.571429                  [Gm7759]   \n",
       "1                14             0   0.000000                    [slck]   \n",
       "2                17             4   0.235294  [slck, Gm31136, Gm24064]   \n",
       "3                13             1   0.076923                       NaN   \n",
       "4                 7             2   0.285714                       NaN   \n",
       "..              ...           ...        ...                       ...   \n",
       "532               3             2   0.666667             [Stol, Tex11]   \n",
       "533               3             2   0.666667                       NaN   \n",
       "534               2             2   1.000000             [Aifm1, Fnld]   \n",
       "535               2             2   1.000000                    [Stol]   \n",
       "536               2             2   1.000000                    [Stol]   \n",
       "\n",
       "       class chrom  \n",
       "0       case  chr1  \n",
       "1       case  chr1  \n",
       "2       case  chr1  \n",
       "3       case  chr1  \n",
       "4       case  chr1  \n",
       "..       ...   ...  \n",
       "532     case  chrX  \n",
       "533     case  chrX  \n",
       "534     case  chrX  \n",
       "535  control  chrX  \n",
       "536  control  chrY  \n",
       "\n",
       "[537 rows x 11 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# This by default is filtered for only genes by default\n",
    "display(all_sg_annot_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>type_index</th>\n",
       "      <th>marker_pos_start</th>\n",
       "      <th>marker_pos_end</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>marker_name</th>\n",
       "      <th>marker_type</th>\n",
       "      <th>marker_feature_type</th>\n",
       "      <th>marker_annot_index</th>\n",
       "      <th>class</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>target</td>\n",
       "      <td>4</td>\n",
       "      <td>13400224.0</td>\n",
       "      <td>40239160.0</td>\n",
       "      <td>Ftms4</td>\n",
       "      <td>fat mass 4</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>424</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>target</td>\n",
       "      <td>4</td>\n",
       "      <td>12916920.0</td>\n",
       "      <td>26517685.0</td>\n",
       "      <td>Fecq3</td>\n",
       "      <td>fecundity QTL 3</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>879</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>target</td>\n",
       "      <td>4</td>\n",
       "      <td>3267623.0</td>\n",
       "      <td>32550581.0</td>\n",
       "      <td>Ebas2</td>\n",
       "      <td>epidermolysis bullosa acquisita severity 2</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>3915</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>target</td>\n",
       "      <td>4</td>\n",
       "      <td>22969802.0</td>\n",
       "      <td>22971000.0</td>\n",
       "      <td>Rr33131</td>\n",
       "      <td>regulatory region 33131</td>\n",
       "      <td>Other Genome Feature</td>\n",
       "      <td>enhancer</td>\n",
       "      <td>21658</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>target</td>\n",
       "      <td>4</td>\n",
       "      <td>22927202.0</td>\n",
       "      <td>22927799.0</td>\n",
       "      <td>Rr33130</td>\n",
       "      <td>regulatory region 33130</td>\n",
       "      <td>Other Genome Feature</td>\n",
       "      <td>enhancer</td>\n",
       "      <td>21669</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14027</th>\n",
       "      <td>target</td>\n",
       "      <td>17</td>\n",
       "      <td>10289113.0</td>\n",
       "      <td>91798267.0</td>\n",
       "      <td>Opfdc9</td>\n",
       "      <td>open field, distance traveled, center 9</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>621954</td>\n",
       "      <td>control</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14028</th>\n",
       "      <td>target</td>\n",
       "      <td>17</td>\n",
       "      <td>61343606.0</td>\n",
       "      <td>89843606.0</td>\n",
       "      <td>Mlh1fc3</td>\n",
       "      <td>MLH1 foci count 3</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>622131</td>\n",
       "      <td>control</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14029</th>\n",
       "      <td>target</td>\n",
       "      <td>17</td>\n",
       "      <td>7226295.0</td>\n",
       "      <td>150107038.0</td>\n",
       "      <td>Dbts2</td>\n",
       "      <td>diabetes 2</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>631602</td>\n",
       "      <td>control</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14030</th>\n",
       "      <td>target</td>\n",
       "      <td>17</td>\n",
       "      <td>54045360.0</td>\n",
       "      <td>148782996.0</td>\n",
       "      <td>Ferq1</td>\n",
       "      <td>genetic fertility QTL 1</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>632342</td>\n",
       "      <td>control</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14031</th>\n",
       "      <td>target</td>\n",
       "      <td>17</td>\n",
       "      <td>47179121.0</td>\n",
       "      <td>129764756.0</td>\n",
       "      <td>Gct6</td>\n",
       "      <td>granulosa cell tumorigenesis 6</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>632436</td>\n",
       "      <td>control</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14032 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         type  type_index marker_pos_start marker_pos_end marker_symbol  \\\n",
       "0      target           4       13400224.0     40239160.0         Ftms4   \n",
       "1      target           4       12916920.0     26517685.0         Fecq3   \n",
       "2      target           4        3267623.0     32550581.0         Ebas2   \n",
       "3      target           4       22969802.0     22971000.0       Rr33131   \n",
       "4      target           4       22927202.0     22927799.0       Rr33130   \n",
       "...       ...         ...              ...            ...           ...   \n",
       "14027  target          17       10289113.0     91798267.0        Opfdc9   \n",
       "14028  target          17       61343606.0     89843606.0       Mlh1fc3   \n",
       "14029  target          17        7226295.0    150107038.0         Dbts2   \n",
       "14030  target          17       54045360.0    148782996.0         Ferq1   \n",
       "14031  target          17       47179121.0    129764756.0          Gct6   \n",
       "\n",
       "                                      marker_name           marker_type  \\\n",
       "0                                      fat mass 4                   QTL   \n",
       "1                                 fecundity QTL 3                   QTL   \n",
       "2      epidermolysis bullosa acquisita severity 2                   QTL   \n",
       "3                         regulatory region 33131  Other Genome Feature   \n",
       "4                         regulatory region 33130  Other Genome Feature   \n",
       "...                                           ...                   ...   \n",
       "14027     open field, distance traveled, center 9                   QTL   \n",
       "14028                           MLH1 foci count 3                   QTL   \n",
       "14029                                  diabetes 2                   QTL   \n",
       "14030                     genetic fertility QTL 1                   QTL   \n",
       "14031              granulosa cell tumorigenesis 6                   QTL   \n",
       "\n",
       "      marker_feature_type marker_annot_index    class chrom  \n",
       "0                     QTL                424     case  chr1  \n",
       "1                     QTL                879     case  chr1  \n",
       "2                     QTL               3915     case  chr1  \n",
       "3                enhancer              21658     case  chr1  \n",
       "4                enhancer              21669     case  chr1  \n",
       "...                   ...                ...      ...   ...  \n",
       "14027                 QTL             621954  control  chrY  \n",
       "14028                 QTL             622131  control  chrY  \n",
       "14029                 QTL             631602  control  chrY  \n",
       "14030                 QTL             632342  control  chrY  \n",
       "14031                 QTL             632436  control  chrY  \n",
       "\n",
       "[14032 rows x 11 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type_index</th>\n",
       "      <th>class</th>\n",
       "      <th>marker_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>case</td>\n",
       "      <td>[nmf172, nmf347, Gn, Gm55427, Hlb290, rsl, Mda...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>control</td>\n",
       "      <td>[rsl, Mdac, Prxl2c, clf2, Vapa, Hdlk, Hdp1, Sn...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>case</td>\n",
       "      <td>[Wif1, Gm48410, Micu2, Gm24794, Map3k4, Gm3162...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>control</td>\n",
       "      <td>[slck, Hycc2, Ndufb3, Bswt, Chpt1, Mybpc1, Gn,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>case</td>\n",
       "      <td>[Gm47343, Gn, Sfi1, Gm33862, 9530037G02Rik, Gm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>335</td>\n",
       "      <td>control</td>\n",
       "      <td>[Ps, Zcchc7, vsd]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>352</td>\n",
       "      <td>case</td>\n",
       "      <td>[Dbf, Gm816]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>425</td>\n",
       "      <td>case</td>\n",
       "      <td>[Dbf]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>430</td>\n",
       "      <td>case</td>\n",
       "      <td>[Tli1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>446</td>\n",
       "      <td>case</td>\n",
       "      <td>[Ps, Nmf191, nmf111]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>173 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     type_index    class                                      marker_symbol\n",
       "0             0     case  [nmf172, nmf347, Gn, Gm55427, Hlb290, rsl, Mda...\n",
       "1             0  control  [rsl, Mdac, Prxl2c, clf2, Vapa, Hdlk, Hdp1, Sn...\n",
       "2             1     case  [Wif1, Gm48410, Micu2, Gm24794, Map3k4, Gm3162...\n",
       "3             1  control  [slck, Hycc2, Ndufb3, Bswt, Chpt1, Mybpc1, Gn,...\n",
       "4             2     case  [Gm47343, Gn, Sfi1, Gm33862, 9530037G02Rik, Gm...\n",
       "..          ...      ...                                                ...\n",
       "168         335  control                                  [Ps, Zcchc7, vsd]\n",
       "169         352     case                                       [Dbf, Gm816]\n",
       "170         425     case                                              [Dbf]\n",
       "171         430     case                                             [Tli1]\n",
       "172         446     case                               [Ps, Nmf191, nmf111]\n",
       "\n",
       "[173 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unique genomic features: 447\n",
      "['Gm7759' 'slck' 'Gm31136' 'Gm24064' 'Rp1' 'Gm37483' 'Sagg' 'Gm29260'\n",
      " 'Smg7' 'Ncf2' 'Fam135a' 'Gm37580' 'Tli1' 'Nek7' 'Ncoa2' 'Dbf' 'Gm816'\n",
      " 'Gm22966' 'Gm54870' 'Gm54871' 'Dock10' 'dde' 'Plcd4' 'Zfp142' 'Sp110'\n",
      " 'Pbx1' 'Arfgef1' 'Cab39' 'Lmbrd1' 'Spag16' 'Gm26245' '4930533P14Rik'\n",
      " 'Dsel' 'Klhl20' 'Ankrd45' 'Tgfb2' 'Gpr39' 'Gm31344' 'Vwc2l' 'Gm38376'\n",
      " 'Hycc2' 'Ndufb3' 'Bswt' 'Aida' 'Gm22681' 'A830018L16Rik' 'Gm37142'\n",
      " 'Vwa3b' 'Gm53552' 'Aff3' 'Aox4' 'Gm29346' 'Gn' 'Tmtc3' 'Xiaf3' 'Dgka'\n",
      " 'nmf172' 'nmf347' 'Wif1' 'Gm48410' 'Gm47343' 'Pcdh15' 'Hsp90b1' 'Gm15344'\n",
      " 'fsq' 'Otogl' 'Igf1' 'Chpt1' 'Mybpc1' 'Timp3' 'Syn3' 'Gm55427' 'Hlb290'\n",
      " 'Sfi1' 'Gm33862' '9530037G02Rik' 'Gm12735' 'Gm11399' 'Arsg' 'Bolt'\n",
      " 'Gm11762' 'Tanc2' 'nmf148' 'Alo1' 'nmf65' 'Asic2' 'Galnt10' 'Nlrp1a'\n",
      " 'nmf420' 'Ism2' 'Scrd2' 'Mta1' 'Eccp' 'Gm48480' 'Gm48479' 'Gm32828'\n",
      " 'Itgb8' 'rsl' 'Mdac' 'Prxl2c' 'clf2' 'Gm41050' 'Elmo1' 'Hecw1' 'Hcn1'\n",
      " '4933413L06Rik' 'Nnt' 'Adarb2' 'Gcnt2' 'Cntnap3' 'Gm28526' 'Gm47665'\n",
      " 'Gm31276' 'Arl15' 'Arid4b' 'Dip2c' 'Hs6st3' 'Gm34061' 'Micu2' 'Dach1'\n",
      " '4930432J09Rik' 'Nek10' 'sky' 'Kcnma1' 'twit' 'Chat' 'Itgbl1' 'Gm49180'\n",
      " 'Tbc1d4' 'Kat6b' 'cmy2' 'Adamts20' 'Ext1' 'Gm34764' 'Celsr1' 'Ttll12'\n",
      " 'Tspo' 'Gm52182' 'uwl' 'Gm19303' 'Parvb' 'Gm3787' 'Gm53900' 'Vps13b'\n",
      " 'Zfpm2' 'Nell2' 'Cal5' 'Cal6' 'Mgrn1' 'Ubald1' 'repro33' 'Robo2' 'Drd3'\n",
      " 'Loh2' 'D530031A16Rik' 'Ccdc80' 'Setd4' 'Gm49659' 'Gm49658'\n",
      " '2510042H03Rik' 'Vapa' 'Gm35400' 'Pknox1' 'hlb156' 'Hst6' 'Gm24794'\n",
      " 'Map3k4' 'CJ186046Rik' 'Gm55893' 'Gm50455' 'Gm50108' 'Gm50107' 'Gm41516'\n",
      " 'l17J2' 'l17J1' 'Zfp563' 'repro17' 'Nudt3' 'Gm49804' 'Hst4' 'Gm31621'\n",
      " 'Fam210a' 'baw' 'Ark2n' 'Dcc' 'Gm32489' 'Me2' 'dblr' 'Papss2' 'stn'\n",
      " 'Gm33649' 'Trpm3' 'ocd' 'Capn1' 'Slc22a30' 'Atad1' 'Mir6715' 'Tectb'\n",
      " 'Gm34116' 'Slit1' 'Nmf220' 'Gm10801' 'Gm10800' 'Tgif2' '4930518I15Rik'\n",
      " '5730471H19Rik' '5430405H02Rik' 'Gm37790' 'Tspan18' 'Frmd4a' 'Gm38085'\n",
      " 'Stk39' 'Malrd1' 'Dwh' 'Ggta1' 'Gm43761' 'Lrrc4c' '4932414N04Rik' 'Lrp1b'\n",
      " 'Tpx2' 'Etl4' 'Nckap1' 'Gm13189' '1700080N15Rik' 'Gm37814'\n",
      " '4930402F06Rik' 'H13' 'nmf62' 'Gm39957' 'Ddx27' 'Syndig1' 'Cacnb2'\n",
      " 'Usp33' 'Zzz3' 'Gm43857' 'Gm43858' 'Gm43859' 'Gm40148' 'Gm43824' 'am3'\n",
      " 'Gm54187' 'repro26' 'Mos3' 'Kcnmb2' 'hlb258' 'Vav3' 'tnyw' 'gro57' 'hem6'\n",
      " 'Gm30991' 'Gm37240' 'Tnik' '4930431L21Rik' 'Rnf13' 'Gm42967' 'Sum2'\n",
      " 'Mhdadsk6' 'Mhdadsk8' 'Gm40106' 'Gm12474' 'Plppr4' 'Gm25653' 'Ps'\n",
      " 'Nmf191' 'nmf111' 'Gm2560' 'Frmd3' 'Lch2' 'Lch' 'vsd' 'Gm11184' 'Gm11413'\n",
      " 'Snapc3' 'Astn2' 'Panc1' 'Pphc2' 'Zfp985' 'Lingo2' 'Gm12381' 'Aadacl4fm4'\n",
      " 'Zcchc7' 'Gm38450' 'Kazn' 'Ptprd' 'Agbl4' 'Cngb3' 'Gm10306' 'Plaa'\n",
      " 'Ift74' 'Gm57080' 'Caap1' '0610009K14Rik' 'Bach2' 'Gm11751' 'Gm34296'\n",
      " 'ctl' 'Susd1' 'Trit1' 'Asph' 'Whl' 'Camta1' 'Kif17' 'Gm25769'\n",
      " '2310002L09Rik' 'Gm11831' 'Virma' 'Mier1' 'Megf9' 'Gm35379' 'Gm11867'\n",
      " 'Fkbp15' 'En2' 'Nmf15' 'Noc4l' 'Mir7026' 'Gm42595' 'Vng' 'repro14'\n",
      " 'Ep400' 'Ddx51' 'Gm22716' 'rs' 'Gm5106' 'Abcb1b' '7330423F06Rik' 'Cc2d2a'\n",
      " 'Lrrc17' 'Fbxl13' 'Xiaf2' 'Stk32b' 'Gm42242' 'Gm52808' 'Micall2' 'l5H1'\n",
      " 'rslk' 'Cracd' 'Cimip2c' 'Gm40339' 'Gm43660' 'AI839979' 'E130118H10Rik'\n",
      " 'Hdlk' 'Hdp1' 'Sndy1' 'wa1l' 'Aak1' 'Thsd7a' 'Smarcad1' 'ssl' 'Shq1'\n",
      " 'Cntn4' 'Gm36408' 'Gm45193' 'Golt1b' 'Plcz1' 'Cntnap2' 'Gm35875'\n",
      " 'Aldh1l1' 'Gm45079' 'repro13' 'jrc' 'nmf206' 'ru2l' 'Apba2' 'psrt' 'l7R4'\n",
      " 'l7R5' 'fit2' 'Chrna7' '4930551E15Rik' 'Hs3st4' 'Gm34174' 'Pde3b'\n",
      " 'Zfp629' 'D7Ertd443e' 'frzl' 'Gm45998' 'Parva' 'Sbf2' 'Sox6' 'Prt5'\n",
      " 'E130304I02Rik' 'Inpp4b' 'Q' 'Hk' 'Cul4a' 'nr' '4933411D12Rik' 'Pcid2'\n",
      " 'Hand2os1' 'Gfod2' 'Gm46039' 'Unc5d' 'nmf242' 'Gpt2' 'Gm15680'\n",
      " 'A330008L17Rik' 'Gm30107' 'repro46' 'repro36' 'repro28' 'Barx2' 'Fat3'\n",
      " 'Gm47604' 'Unc13c' 'Gm28703' 'Ltw3' 'n-TKctt18' 'Morf4l1' 'Gm48091'\n",
      " 'Opcml' 'Wdr72' 'Tln2' 'avc3' 'Gm35715' 'cir' 'Ccdc12' 'Or8b35' 'Csnk1g1'\n",
      " 'Gfral' 'Phip' 'Plscr2' 'Pknox2' 'Cadm1' 'Cd109' 'Gramd1b' 'Ireb2'\n",
      " 'Gm10719' 'Gm10718' 'Gm10717' 'Gm10720' 'Gm10721' 'Gm10722' 'Gm11168'\n",
      " 'Gm17535' 'Pdgfd' 'Sidt2' 'Coro2b' 'Gm28548' 'Gm31614' 'Stol' 'Fnld' 'Li'\n",
      " 'Ppef1' 'Akap14' 'Tex11' 'Aifm1']\n"
     ]
    }
   ],
   "source": [
    "# these are duplicated sometimes between cases and controls\n",
    "# but that's ok cause we just want the unique list of genes that we should investigate further\n",
    "display(all_sig_annot_df)\n",
    "\n",
    "# What about just looking at only genes?\n",
    "gene_tmp = all_sig_annot_df[all_sig_annot_df[\"marker_type\"] == \"Gene\"]\n",
    "\n",
    "gene_only = gene_tmp.groupby([\"type_index\", \"class\"]).agg(list)[\"marker_symbol\"].reset_index()\n",
    "display(gene_only)\n",
    "\n",
    "uniq_count = gene_tmp[\"marker_symbol\"].unique().shape[0]\n",
    "print(f\"unique genomic features: {uniq_count}\")\n",
    "\n",
    "candidate_genes = gene_tmp[\"marker_symbol\"].unique()\n",
    "print(candidate_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What about breaking up large CISs? It appears they are too large and span multiple genes\n",
    "\n",
    "# Or do now we check each individual gene that was found to further refine our search?\n",
    "\n",
    "# What's stopping me from testing literally every gene with +/-50 kb region of insertions with a ranksums test?\n",
    "# because we don't want this to be gene centric, however, it's not a bad way to validate the candidate gene list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rpl48-ps1 1 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>914</th>\n",
       "      <td>case</td>\n",
       "      <td>chr10</td>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>57513025</td>\n",
       "      <td>57513025</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     type  chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "914  case  chr10        66      1      0            1  57513025  57513025   \n",
       "\n",
       "     range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "914      0                 1             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cpd 0 0\n",
      "Ckb 0 0\n",
      "Ehhadh 1 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1960</th>\n",
       "      <td>case</td>\n",
       "      <td>chr16</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "      <td>325</td>\n",
       "      <td>89</td>\n",
       "      <td>21614194</td>\n",
       "      <td>21614302</td>\n",
       "      <td>108</td>\n",
       "      <td>26</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type  chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "1960  case  chr16         1     26    325           89  21614194  21614302   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "1960    108                26             7                   14  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1629</th>\n",
       "      <td>control</td>\n",
       "      <td>chr16</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>1891</td>\n",
       "      <td>790</td>\n",
       "      <td>21600392</td>\n",
       "      <td>21614302</td>\n",
       "      <td>13910</td>\n",
       "      <td>62</td>\n",
       "      <td>602</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         type  chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "1629  control  chr16         1     62   1891          790  21600392  21614302   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "1629  13910                62           602                   32  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Son 0 0\n",
      "Macrod2 2 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2552</th>\n",
       "      <td>case</td>\n",
       "      <td>chr2</td>\n",
       "      <td>63</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>140313463</td>\n",
       "      <td>140313466</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2648</th>\n",
       "      <td>case</td>\n",
       "      <td>chr2</td>\n",
       "      <td>159</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>140753859</td>\n",
       "      <td>140753859</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type chrom  subgraph  nodes  edges  num_inserts    min_pos    max_pos  \\\n",
       "2552  case  chr2        63      2      1           33  140313463  140313466   \n",
       "2648  case  chr2       159      1      0           15  140753859  140753859   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "2552      3                 2             0                    0  \n",
       "2648      0                 1             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sprr1b 1 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2848</th>\n",
       "      <td>case</td>\n",
       "      <td>chr3</td>\n",
       "      <td>83</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>92347698</td>\n",
       "      <td>92347698</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "2848  case  chr3        83      1      0           41  92347698  92347698   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "2848      0                 1             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2484</th>\n",
       "      <td>control</td>\n",
       "      <td>chr3</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>438</td>\n",
       "      <td>92347697</td>\n",
       "      <td>92347714</td>\n",
       "      <td>17</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         type chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "2484  control  chr3         7      6     15          438  92347697  92347714   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "2484     17                 6             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Eif3b 0 0\n",
      "Aak1 2 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3771</th>\n",
       "      <td>case</td>\n",
       "      <td>chr6</td>\n",
       "      <td>1</td>\n",
       "      <td>75</td>\n",
       "      <td>2775</td>\n",
       "      <td>838</td>\n",
       "      <td>86910715</td>\n",
       "      <td>86910899</td>\n",
       "      <td>184</td>\n",
       "      <td>75</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3885</th>\n",
       "      <td>case</td>\n",
       "      <td>chr6</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>86976664</td>\n",
       "      <td>86976664</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "3771  case  chr6         1     75   2775          838  86910715  86910899   \n",
       "3885  case  chr6       115      1      0           55  86976664  86976664   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "3771    184                75             4                   19  \n",
       "3885      0                 1             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3295</th>\n",
       "      <td>control</td>\n",
       "      <td>chr6</td>\n",
       "      <td>1</td>\n",
       "      <td>114</td>\n",
       "      <td>6441</td>\n",
       "      <td>10434</td>\n",
       "      <td>86910504</td>\n",
       "      <td>86910899</td>\n",
       "      <td>395</td>\n",
       "      <td>114</td>\n",
       "      <td>8</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3329</th>\n",
       "      <td>control</td>\n",
       "      <td>chr6</td>\n",
       "      <td>35</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>86858016</td>\n",
       "      <td>86858023</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         type chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "3295  control  chr6         1    114   6441        10434  86910504  86910899   \n",
       "3329  control  chr6        35      3      3           16  86858016  86858023   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "3295    395               114             8                   26  \n",
       "3329      7                 3             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rnf214 0 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3965</th>\n",
       "      <td>control</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>36</td>\n",
       "      <td>774</td>\n",
       "      <td>45852161</td>\n",
       "      <td>45852171</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         type chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "3965  control  chr9         4      9     36          774  45852161  45852171   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "3965     10                 9             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# TODO: use list of Laura genes and see why that genomic region did not show up\n",
    "top10genes = [\"Aak1\", \"Ehhadh\", \"Macrod2\", \"Ckb\", \"Rnf214\", \"Sprr1b\", \"Cpd\", \"Rpl48-ps1\", \"Son\", \"Eif3b\"]\n",
    "top10annot = annot_df[annot_df[\"Marker Symbol\"].isin(top10genes)]\n",
    "for row in top10annot.itertuples():\n",
    "    chrom = f\"chr{row.Chr}\"\n",
    "    case_chrom_df = case_df[case_df[\"chrom\"] == chrom]\n",
    "    control_chrom_df = control_df[control_df[\"chrom\"] == chrom]\n",
    "    res1 = int(row._4 - 50000) <= case_chrom_df[\"max_pos\"]\n",
    "    res2 = int(row._5 + 50000) >= case_chrom_df[\"min_pos\"]\n",
    "    res3 = int(row._4 - 50000) <= control_chrom_df[\"max_pos\"]\n",
    "    res4 = int(row._5 + 50000) >= control_chrom_df[\"min_pos\"]\n",
    "\n",
    "    print(row._7, (res1 & res2).sum(), (res3 & res4).sum())\n",
    "    if (res1 & res2).sum() != 0:\n",
    "        display(case_chrom_df.loc[res1 & res2])\n",
    "    if (res3 & res4).sum() != 0:\n",
    "        display(control_chrom_df.loc[res3 & res4])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: check for each gene with an extended proomoter the case vs controls insertions\n",
    "# check with ranksums test as well as binomial test\n",
    "# What does the candidate gene list look like now?\n",
    "\n",
    "\n",
    "# TODO: I need to go through my code again. Go step by step for each function in another notebook\n",
    "# double check this code and cis_networks.py\n",
    "\n",
    "\n",
    "# TODO: rerun without mapq thresholding in preprocess_reads and preprocess_insertions\n",
    "\n",
    "\n",
    "# TODO: should we be using normalized read counts? idk, what stage would this occur at\n",
    "# the variation would come from read depth, so we would have to normalize based on this BEFORE preprocess_reads.py\n",
    "# What is done with RNAseq? How to normalize on read depth while still keeping count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# volcano plot of p-values and LFC for each CIS if wanted\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output list of pCIS (now can be called CIS) that are sig.\n",
    "# per CIS, output list of TAs that are sig.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# take identified candidate CIS/genes and check for co-occurrence with all other CIS/genes\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "netcis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
