{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn.objects as so\n",
    "from seaborn import axes_style\n",
    "import networkx as nx\n",
    "from scipy.stats import chisquare, binomtest, ranksums, mannwhitneyu, skewtest, kurtosistest\n",
    "\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'netcis.network_analysis' from '/project/cs-myers/MathewF/projects/Laura-SB-Analysis/NetCIS/netcis/network_analysis.py'>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys, os\n",
    "dir1 = os.path.dirname(os.path.abspath(''))\n",
    "if not dir1 in sys.path: sys.path.append(dir1)\n",
    "from netcis import network_analysis as na\n",
    "\n",
    "from importlib import reload\n",
    "reload(na)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_dir = Path(\"/project/cs-myers/MathewF/projects/Laura-SB-Analysis/output/2020_SB-no_threshold-graphs/\")\n",
    "ta_dir = Path(\"/project/cs-myers/MathewF/software/bowtie2-2.4.5/indexes/GRCm39_TAs/\")\n",
    "gene_annot = Path(\"/project/cs-myers/MathewF/projects/Laura-SB-Analysis/NetCIS/toy-data/MRK_List2.rpt\")\n",
    "output_prefix = \"/project/cs-myers/MathewF/projects/Laura-SB-Analysis/output/2020_SB-no_threshold\"\n",
    "output = Path(output_prefix + \"-analysis\")\n",
    "output.mkdir(exist_ok=True)\n",
    "ta_error = 5\n",
    "pval_threshold = 0.05\n",
    "verbose = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "annot_df = pd.read_csv(gene_annot, sep=\"\\t\")\n",
    "annot_df = annot_df[pd.notna(annot_df[\"genome coordinate start\"])].drop(\"Status\", axis=1)\n",
    "annot_df[\"chrom\"] = annot_df[\"Chr\"].apply(lambda x: f\"chr{x}\")\n",
    "annot_df = annot_df.sort_values([\"chrom\"]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_files = { file.name.split(\".\")[0]: pd.read_csv(file, sep=\"\\t\", header=None) for file in ta_dir.iterdir() }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_subgraph_dict = na.get_subgraphs(graph_dir, \"case\")\n",
    "case_df = na.get_subgraph_stats(case_subgraph_dict, \"case\", bed_files, ta_error)\n",
    "case_df.sort_values([\"chrom\", \"subgraph\", \"nodes\"]).to_csv(output / \"case_df.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "control_subgraph_dict = na.get_subgraphs(graph_dir, \"control\")\n",
    "control_df = na.get_subgraph_stats(control_subgraph_dict, \"control\", bed_files, ta_error)\n",
    "control_df.sort_values([\"chrom\", \"subgraph\", \"nodes\"]).to_csv(output / \"control_df.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "\tsig. genomic features: 749/42646\n",
      "chr10\n",
      "\tsig. genomic features: 219/31345\n",
      "chr11\n",
      "\tsig. genomic features: 235/43002\n",
      "chr12\n",
      "\tsig. genomic features: 83/26305\n",
      "chr13\n",
      "\tsig. genomic features: 154/27803\n",
      "chr14\n",
      "\tsig. genomic features: 64/25866\n",
      "chr15\n",
      "\tsig. genomic features: 181/25463\n",
      "chr16\n",
      "\tsig. genomic features: 110/21683\n",
      "chr17\n",
      "\tsig. genomic features: 186/27635\n",
      "chr18\n",
      "\tsig. genomic features: 57/19494\n",
      "chr19\n",
      "\tsig. genomic features: 81/19099\n",
      "chr2\n",
      "\tsig. genomic features: 356/48037\n",
      "chr3\n",
      "\tsig. genomic features: 147/32767\n",
      "chr4\n",
      "\tsig. genomic features: 663/38690\n",
      "chr5\n",
      "\tsig. genomic features: 261/41290\n",
      "chr6\n",
      "\tsig. genomic features: 233/36077\n",
      "chr7\n",
      "\tsig. genomic features: 189/41139\n",
      "chr8\n",
      "\tsig. genomic features: 200/31828\n",
      "chr9\n",
      "\tsig. genomic features: 341/35025\n",
      "chrM\n",
      "chrX\n",
      "\tsig. genomic features: 100/17078\n",
      "chrY\n"
     ]
    }
   ],
   "source": [
    "reload(na)\n",
    "chroms = case_df[\"chrom\"].sort_values().unique()\n",
    "all_features_list = []\n",
    "# TODO: parallelize this?\n",
    "for chrom in chroms:\n",
    "    # if chrom != \"chrM\":\n",
    "    #     continue\n",
    "    print(chrom)\n",
    "    \n",
    "    # get chromosome subsets for case and control\n",
    "    case_chrom_df = case_df[case_df[\"chrom\"] == chrom]\n",
    "    control_chrom_df = control_df[control_df[\"chrom\"] == chrom]\n",
    "    case_chrom_subgraphs = case_subgraph_dict[chrom]\n",
    "    control_chrom_subgraphs = control_subgraph_dict[chrom]\n",
    "    annot_chrom_df = annot_df[annot_df[\"chrom\"] == chrom]\n",
    "    \n",
    "    # cases as the target\n",
    "    case_overlaps = na.pcis_overlaps(case_chrom_df, control_chrom_df)\n",
    "    if not case_overlaps:  # if empty\n",
    "        case_genes = None\n",
    "    else:\n",
    "        case_TA_df, case_overall_df = na.compare_pcis(case_overlaps, case_chrom_subgraphs, control_chrom_subgraphs)\n",
    "        case_sig_df = na.pcis_to_cis(case_overall_df, pval_threshold)\n",
    "        if len(case_sig_df) != 0:\n",
    "            case_genes = na.cis_annotate(case_sig_df, annot_chrom_df)\n",
    "            case_genes[\"class\"] = \"case\"\n",
    "        else:\n",
    "            case_genes = None\n",
    "    \n",
    "    # controls as the target\n",
    "    control_overlaps = na.pcis_overlaps(control_chrom_df, case_chrom_df)\n",
    "    if not control_overlaps:  # if empty\n",
    "        control_genes = None\n",
    "    else:\n",
    "        control_TA_df, control_overall_df = na.compare_pcis(control_overlaps, control_chrom_subgraphs, case_chrom_subgraphs)\n",
    "        control_sig_df = na.pcis_to_cis(control_overall_df, pval_threshold)\n",
    "        if len(control_sig_df) != 0:\n",
    "            control_genes = na.cis_annotate(control_sig_df, annot_chrom_df)\n",
    "            control_genes[\"class\"] = \"control\"\n",
    "        else:\n",
    "            control_genes = None\n",
    "    \n",
    "    if case_genes is not None and control_genes is not None:\n",
    "        both_genes = pd.concat([case_genes, control_genes], ignore_index=True)\n",
    "    elif case_genes is not None:\n",
    "        both_genes = case_genes\n",
    "    elif control_genes is not None:\n",
    "        both_genes = control_genes\n",
    "    else:  # both are none\n",
    "        print(\"no significant genomic features found\")\n",
    "        continue\n",
    "    \n",
    "    both_genes[\"chrom\"] = chrom\n",
    "    all_features_list.append(both_genes)\n",
    "    if verbose > 0:\n",
    "        # print(len(both_genes))\n",
    "        print(len(both_genes[\"marker_symbol\"].unique()))\n",
    "\n",
    "    # TODO: are there too many repeated genes? Does this make sense that they would be repeated?\n",
    "    # it appears that sometimes there are multiple CIS in a gene, because the CIS range is quite small\n",
    "    print(f\"\"\"\\tsig. genomic features: {both_genes[\"marker_symbol\"].unique().shape[0]}/{annot_chrom_df[\"Marker Symbol\"].unique().shape[0]}\"\"\")\n",
    "\n",
    "# get all genomic features\n",
    "all_features_df = pd.concat(all_features_list, ignore_index=True)\n",
    "# all_genes_df.to_csv(output / \"all_genes.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>type_index</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>marker_name</th>\n",
       "      <th>marker_type</th>\n",
       "      <th>marker_feature_type</th>\n",
       "      <th>marker_annot_index</th>\n",
       "      <th>class</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>target</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Feml5</td>\n",
       "      <td>femur length 5, 5 week</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>755</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>target</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Fearet1</td>\n",
       "      <td>fear retrieval 1</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>967</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>target</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Rr12852</td>\n",
       "      <td>regulatory region 12852</td>\n",
       "      <td>Other Genome Feature</td>\n",
       "      <td>CTCF binding site</td>\n",
       "      <td>1037</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>target</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Gcsfis</td>\n",
       "      <td>G-CSF induced splenomegaly</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>1693</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>target</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Eila1</td>\n",
       "      <td>ethanol induced locomotor activity 1</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>3147</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20514</th>\n",
       "      <td>target</td>\n",
       "      <td>130.0</td>\n",
       "      <td>Femd9</td>\n",
       "      <td>femur midshaft diameter 9, 10 week</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>632357</td>\n",
       "      <td>control</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20515</th>\n",
       "      <td>target</td>\n",
       "      <td>130.0</td>\n",
       "      <td>Femd12</td>\n",
       "      <td>femur midshaft diameter 12, 16 week</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>632358</td>\n",
       "      <td>control</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20516</th>\n",
       "      <td>target</td>\n",
       "      <td>130.0</td>\n",
       "      <td>Fnld</td>\n",
       "      <td>faint-lined</td>\n",
       "      <td>Gene</td>\n",
       "      <td>heritable phenotypic marker</td>\n",
       "      <td>632402</td>\n",
       "      <td>control</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20517</th>\n",
       "      <td>target</td>\n",
       "      <td>130.0</td>\n",
       "      <td>Gct6</td>\n",
       "      <td>granulosa cell tumorigenesis 6</td>\n",
       "      <td>QTL</td>\n",
       "      <td>QTL</td>\n",
       "      <td>632436</td>\n",
       "      <td>control</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20518</th>\n",
       "      <td>target</td>\n",
       "      <td>167.0</td>\n",
       "      <td>Gm51614</td>\n",
       "      <td>predicted gene, 51614</td>\n",
       "      <td>Gene</td>\n",
       "      <td>lncRNA gene</td>\n",
       "      <td>621130</td>\n",
       "      <td>control</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20519 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         type  type_index marker_symbol                           marker_name  \\\n",
       "0      target         2.0         Feml5                femur length 5, 5 week   \n",
       "1      target         2.0       Fearet1                      fear retrieval 1   \n",
       "2      target         2.0       Rr12852               regulatory region 12852   \n",
       "3      target         2.0        Gcsfis            G-CSF induced splenomegaly   \n",
       "4      target         2.0         Eila1  ethanol induced locomotor activity 1   \n",
       "...       ...         ...           ...                                   ...   \n",
       "20514  target       130.0         Femd9    femur midshaft diameter 9, 10 week   \n",
       "20515  target       130.0        Femd12   femur midshaft diameter 12, 16 week   \n",
       "20516  target       130.0          Fnld                           faint-lined   \n",
       "20517  target       130.0          Gct6        granulosa cell tumorigenesis 6   \n",
       "20518  target       167.0       Gm51614                 predicted gene, 51614   \n",
       "\n",
       "                marker_type          marker_feature_type marker_annot_index  \\\n",
       "0                       QTL                          QTL                755   \n",
       "1                       QTL                          QTL                967   \n",
       "2      Other Genome Feature            CTCF binding site               1037   \n",
       "3                       QTL                          QTL               1693   \n",
       "4                       QTL                          QTL               3147   \n",
       "...                     ...                          ...                ...   \n",
       "20514                   QTL                          QTL             632357   \n",
       "20515                   QTL                          QTL             632358   \n",
       "20516                  Gene  heritable phenotypic marker             632402   \n",
       "20517                   QTL                          QTL             632436   \n",
       "20518                  Gene                  lncRNA gene             621130   \n",
       "\n",
       "         class chrom  \n",
       "0         case  chr1  \n",
       "1         case  chr1  \n",
       "2         case  chr1  \n",
       "3         case  chr1  \n",
       "4         case  chr1  \n",
       "...        ...   ...  \n",
       "20514  control  chrX  \n",
       "20515  control  chrX  \n",
       "20516  control  chrX  \n",
       "20517  control  chrX  \n",
       "20518  control  chrX  \n",
       "\n",
       "[20519 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(all_features_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>type_index</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>marker_name</th>\n",
       "      <th>marker_type</th>\n",
       "      <th>marker_feature_type</th>\n",
       "      <th>marker_annot_index</th>\n",
       "      <th>class</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>target</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Dnah7a</td>\n",
       "      <td>dynein, axonemal, heavy chain 7A</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>2192</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>target</td>\n",
       "      <td>15.0</td>\n",
       "      <td>Fam135a</td>\n",
       "      <td>family with sequence similarity 135, member A</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>726</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>target</td>\n",
       "      <td>64.0</td>\n",
       "      <td>Cdh7</td>\n",
       "      <td>cadherin 7, type 2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>37082</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>target</td>\n",
       "      <td>109.0</td>\n",
       "      <td>Myo1b</td>\n",
       "      <td>myosin IB</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>21759</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>target</td>\n",
       "      <td>153.0</td>\n",
       "      <td>Ncoa2</td>\n",
       "      <td>nuclear receptor coactivator 2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>20423</td>\n",
       "      <td>case</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20386</th>\n",
       "      <td>target</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Ppef1</td>\n",
       "      <td>protein phosphatase with EF hand calcium-bindi...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>622696</td>\n",
       "      <td>control</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20395</th>\n",
       "      <td>target</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Il1rapl2</td>\n",
       "      <td>interleukin 1 receptor accessory protein-like 2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>621712</td>\n",
       "      <td>control</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20416</th>\n",
       "      <td>target</td>\n",
       "      <td>32.0</td>\n",
       "      <td>Akap14</td>\n",
       "      <td>A kinase anchor protein 14</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>616573</td>\n",
       "      <td>control</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20447</th>\n",
       "      <td>target</td>\n",
       "      <td>64.0</td>\n",
       "      <td>Tex11</td>\n",
       "      <td>testis expressed gene 11</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>620727</td>\n",
       "      <td>control</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20505</th>\n",
       "      <td>target</td>\n",
       "      <td>130.0</td>\n",
       "      <td>Aifm1</td>\n",
       "      <td>apoptosis-inducing factor, mitochondrion-assoc...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>616569</td>\n",
       "      <td>control</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>424 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         type  type_index marker_symbol  \\\n",
       "45     target         7.0        Dnah7a   \n",
       "175    target        15.0       Fam135a   \n",
       "436    target        64.0          Cdh7   \n",
       "596    target       109.0         Myo1b   \n",
       "718    target       153.0         Ncoa2   \n",
       "...       ...         ...           ...   \n",
       "20386  target        24.0         Ppef1   \n",
       "20395  target        30.0      Il1rapl2   \n",
       "20416  target        32.0        Akap14   \n",
       "20447  target        64.0         Tex11   \n",
       "20505  target       130.0         Aifm1   \n",
       "\n",
       "                                             marker_name marker_type  \\\n",
       "45                      dynein, axonemal, heavy chain 7A        Gene   \n",
       "175        family with sequence similarity 135, member A        Gene   \n",
       "436                                   cadherin 7, type 2        Gene   \n",
       "596                                            myosin IB        Gene   \n",
       "718                       nuclear receptor coactivator 2        Gene   \n",
       "...                                                  ...         ...   \n",
       "20386  protein phosphatase with EF hand calcium-bindi...        Gene   \n",
       "20395    interleukin 1 receptor accessory protein-like 2        Gene   \n",
       "20416                         A kinase anchor protein 14        Gene   \n",
       "20447                           testis expressed gene 11        Gene   \n",
       "20505  apoptosis-inducing factor, mitochondrion-assoc...        Gene   \n",
       "\n",
       "       marker_feature_type marker_annot_index    class chrom  \n",
       "45     protein coding gene               2192     case  chr1  \n",
       "175    protein coding gene                726     case  chr1  \n",
       "436    protein coding gene              37082     case  chr1  \n",
       "596    protein coding gene              21759     case  chr1  \n",
       "718    protein coding gene              20423     case  chr1  \n",
       "...                    ...                ...      ...   ...  \n",
       "20386  protein coding gene             622696  control  chrX  \n",
       "20395  protein coding gene             621712  control  chrX  \n",
       "20416  protein coding gene             616573  control  chrX  \n",
       "20447  protein coding gene             620727  control  chrX  \n",
       "20505  protein coding gene             616569  control  chrX  \n",
       "\n",
       "[424 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene_tmp = all_features_df[(all_features_df[\"marker_type\"] == \"Gene\") & (all_features_df[\"marker_feature_type\"] == \"protein coding gene\")]\n",
    "display(gene_tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type_index</th>\n",
       "      <th>class</th>\n",
       "      <th>marker_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>case</td>\n",
       "      <td>[Sfi1, Prxl2c, Vapa, Gm10801, Gm10800, Rab5if,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>control</td>\n",
       "      <td>[Sp140l2, Gm53567, Prxl2c, Vapa, Gm10801, Gm10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>case</td>\n",
       "      <td>[Adamts20, Aak1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>control</td>\n",
       "      <td>[Sfi1, Ppard, Trpm3, Aak1, Zfp954, Zfp773]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.0</td>\n",
       "      <td>case</td>\n",
       "      <td>[Chpt1, Mybpc1, Hs6st3, Zfpm2, Tspan18, Barx2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>351.0</td>\n",
       "      <td>control</td>\n",
       "      <td>[Kif17]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>357.0</td>\n",
       "      <td>case</td>\n",
       "      <td>[Astn2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>371.0</td>\n",
       "      <td>control</td>\n",
       "      <td>[Zcchc7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>446.0</td>\n",
       "      <td>case</td>\n",
       "      <td>[Zcchc7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>518.0</td>\n",
       "      <td>control</td>\n",
       "      <td>[Rpe]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>174 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     type_index    class                                      marker_symbol\n",
       "0           0.0     case  [Sfi1, Prxl2c, Vapa, Gm10801, Gm10800, Rab5if,...\n",
       "1           0.0  control  [Sp140l2, Gm53567, Prxl2c, Vapa, Gm10801, Gm10...\n",
       "2           1.0     case                                   [Adamts20, Aak1]\n",
       "3           1.0  control         [Sfi1, Ppard, Trpm3, Aak1, Zfp954, Zfp773]\n",
       "4           2.0     case     [Chpt1, Mybpc1, Hs6st3, Zfpm2, Tspan18, Barx2]\n",
       "..          ...      ...                                                ...\n",
       "169       351.0  control                                            [Kif17]\n",
       "170       357.0     case                                            [Astn2]\n",
       "171       371.0  control                                           [Zcchc7]\n",
       "172       446.0     case                                           [Zcchc7]\n",
       "173       518.0  control                                              [Rpe]\n",
       "\n",
       "[174 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene_only = gene_tmp.groupby([\"type_index\", \"class\"]).agg(list)[\"marker_symbol\"].reset_index()\n",
    "display(gene_only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unique genomic features: 298\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['2310002L09Rik',\n",
       " '4930402F06Rik',\n",
       " '4932414N04Rik',\n",
       " 'A830018L16Rik',\n",
       " 'Aadacl4fm4',\n",
       " 'Aak1',\n",
       " 'Abcb1b',\n",
       " 'Acot12',\n",
       " 'Acvr2a',\n",
       " 'Adamts20',\n",
       " 'Adarb2',\n",
       " 'Aff3',\n",
       " 'Agbl4',\n",
       " 'Agps',\n",
       " 'Aida',\n",
       " 'Aifm1',\n",
       " 'Akap14',\n",
       " 'Aldh1l1',\n",
       " 'Alk',\n",
       " 'Ankrd45',\n",
       " 'Aox4',\n",
       " 'Apba2',\n",
       " 'Arfgef1',\n",
       " 'Arid4b',\n",
       " 'Ark2n',\n",
       " 'Arsg',\n",
       " 'Asic2',\n",
       " 'Asph',\n",
       " 'Astn2',\n",
       " 'Atad1',\n",
       " 'Atic',\n",
       " 'BC147527',\n",
       " 'Bach2',\n",
       " 'Barx2',\n",
       " 'Caap1',\n",
       " 'Cab39',\n",
       " 'Cacnb2',\n",
       " 'Cadm1',\n",
       " 'Camta1',\n",
       " 'Capn1',\n",
       " 'Cask',\n",
       " 'Cc2d2a',\n",
       " 'Ccdc12',\n",
       " 'Ccdc80',\n",
       " 'Cd109',\n",
       " 'Cdh7',\n",
       " 'Cdk5rap2',\n",
       " 'Celsr1',\n",
       " 'Cfhr3',\n",
       " 'Chat',\n",
       " 'Chl1',\n",
       " 'Chpt1',\n",
       " 'Chrna7',\n",
       " 'Cimip2c',\n",
       " 'Cngb3',\n",
       " 'Cntn4',\n",
       " 'Cntnap2',\n",
       " 'Cntnap3',\n",
       " 'Coro2b',\n",
       " 'Cracd',\n",
       " 'Csnk1g1',\n",
       " 'Cul4a',\n",
       " 'D7Ertd443e',\n",
       " 'Dab1',\n",
       " 'Dach1',\n",
       " 'Dcc',\n",
       " 'Ddx27',\n",
       " 'Ddx51',\n",
       " 'Dgka',\n",
       " 'Dip2c',\n",
       " 'Dnah7a',\n",
       " 'Dock10',\n",
       " 'Dop1a',\n",
       " 'Drd3',\n",
       " 'Dsel',\n",
       " 'Elmo1',\n",
       " 'En2',\n",
       " 'Ep400',\n",
       " 'Etl4',\n",
       " 'Ext1',\n",
       " 'Fam135a',\n",
       " 'Fam172a',\n",
       " 'Fam210a',\n",
       " 'Fbxl13',\n",
       " 'Fgfr4',\n",
       " 'Fkbp15',\n",
       " 'Fn1',\n",
       " 'Frmd3',\n",
       " 'Frmd4a',\n",
       " 'Galnt10',\n",
       " 'Gcnt2',\n",
       " 'Gfod2',\n",
       " 'Gfral',\n",
       " 'Ggta1',\n",
       " 'Gm10306',\n",
       " 'Gm10715',\n",
       " 'Gm10717',\n",
       " 'Gm10718',\n",
       " 'Gm10719',\n",
       " 'Gm10720',\n",
       " 'Gm10721',\n",
       " 'Gm10722',\n",
       " 'Gm10800',\n",
       " 'Gm10801',\n",
       " 'Gm11168',\n",
       " 'Gm11232',\n",
       " 'Gm11756',\n",
       " 'Gm11757',\n",
       " 'Gm13871',\n",
       " 'Gm15085',\n",
       " 'Gm17535',\n",
       " 'Gm20385',\n",
       " 'Gm29758',\n",
       " 'Gm34296',\n",
       " 'Gm37240',\n",
       " 'Gm38510',\n",
       " 'Gm4593',\n",
       " 'Gm49804',\n",
       " 'Gm53253',\n",
       " 'Gm53268',\n",
       " 'Gm53405',\n",
       " 'Gm53567',\n",
       " 'Gm7247',\n",
       " 'Gm7609',\n",
       " 'Gm9758',\n",
       " 'Gml',\n",
       " 'Gml2',\n",
       " 'Golt1b',\n",
       " 'Gpr39',\n",
       " 'Gpt2',\n",
       " 'Gramd1b',\n",
       " 'H13',\n",
       " 'Hcn1',\n",
       " 'Hecw1',\n",
       " 'Hs3st4',\n",
       " 'Hs6st3',\n",
       " 'Hsp90b1',\n",
       " 'Hycc2',\n",
       " 'Ift74',\n",
       " 'Igf1',\n",
       " 'Il1rapl2',\n",
       " 'Inpp4b',\n",
       " 'Ireb2',\n",
       " 'Ism2',\n",
       " 'Itgb8',\n",
       " 'Itgbl1',\n",
       " 'Kat6b',\n",
       " 'Kazn',\n",
       " 'Kcnd2',\n",
       " 'Kcnma1',\n",
       " 'Kcnmb2',\n",
       " 'Kif17',\n",
       " 'Klhl20',\n",
       " 'Lingo2',\n",
       " 'Lmbrd1',\n",
       " 'Lrp1b',\n",
       " 'Lrrc17',\n",
       " 'Lrrc4c',\n",
       " 'Malrd1',\n",
       " 'Map3k4',\n",
       " 'Me2',\n",
       " 'Megf9',\n",
       " 'Mgrn1',\n",
       " 'Micall2',\n",
       " 'Micu2',\n",
       " 'Mier1',\n",
       " 'Mms22l',\n",
       " 'Morf4l1',\n",
       " 'Mrc1',\n",
       " 'Mroh9',\n",
       " 'Mta1',\n",
       " 'Mup10',\n",
       " 'Mup12',\n",
       " 'Mup16',\n",
       " 'Mup17',\n",
       " 'Mup19',\n",
       " 'Mup22',\n",
       " 'Mybpc1',\n",
       " 'Myo1b',\n",
       " 'Ncf2',\n",
       " 'Nckap1',\n",
       " 'Ncoa2',\n",
       " 'Ndufb3',\n",
       " 'Nedd9',\n",
       " 'Nek10',\n",
       " 'Nek7',\n",
       " 'Nell2',\n",
       " 'Nlrp1a',\n",
       " 'Nnt',\n",
       " 'Noc4l',\n",
       " 'Nsun7',\n",
       " 'Nudt3',\n",
       " 'Opcml',\n",
       " 'Or10p21',\n",
       " 'Or10p22',\n",
       " 'Or8b35',\n",
       " 'Orc4',\n",
       " 'Otogl',\n",
       " 'Papss2',\n",
       " 'Parva',\n",
       " 'Parvb',\n",
       " 'Pbx1',\n",
       " 'Pcdh15',\n",
       " 'Pcid2',\n",
       " 'Pde3b',\n",
       " 'Pdgfd',\n",
       " 'Phip',\n",
       " 'Pkhd1',\n",
       " 'Pknox1',\n",
       " 'Pknox2',\n",
       " 'Plaa',\n",
       " 'Plcd4',\n",
       " 'Plcz1',\n",
       " 'Plppr4',\n",
       " 'Plscr2',\n",
       " 'Ppard',\n",
       " 'Ppef1',\n",
       " 'Pramel59',\n",
       " 'Prkca',\n",
       " 'Prxl2c',\n",
       " 'Ptprd',\n",
       " 'Rab5if',\n",
       " 'Rnf13',\n",
       " 'Robo2',\n",
       " 'Rp1',\n",
       " 'Rpe',\n",
       " 'Sbf2',\n",
       " 'Setd4',\n",
       " 'Sfi1',\n",
       " 'Shq1',\n",
       " 'Sidt2',\n",
       " 'Sla2',\n",
       " 'Slc22a30',\n",
       " 'Slit1',\n",
       " 'Smarcad1',\n",
       " 'Smg7',\n",
       " 'Snapc3',\n",
       " 'Sox6',\n",
       " 'Sp140l1',\n",
       " 'Sp140l2',\n",
       " 'Spag16',\n",
       " 'Speer4e',\n",
       " 'Stk32b',\n",
       " 'Stk39',\n",
       " 'Susd1',\n",
       " 'Syn3',\n",
       " 'Syndig1',\n",
       " 'Tanc2',\n",
       " 'Tbc1d4',\n",
       " 'Tcstv3',\n",
       " 'Tectb',\n",
       " 'Tex11',\n",
       " 'Tgfb2',\n",
       " 'Tgif2',\n",
       " 'Thsd7a',\n",
       " 'Timp3',\n",
       " 'Tln2',\n",
       " 'Tmtc3',\n",
       " 'Tnik',\n",
       " 'Tox4',\n",
       " 'Tpx2',\n",
       " 'Trit1',\n",
       " 'Trpm3',\n",
       " 'Tspan18',\n",
       " 'Tspo',\n",
       " 'Ttll12',\n",
       " 'Ubald1',\n",
       " 'Ube3d',\n",
       " 'Unc13c',\n",
       " 'Unc5d',\n",
       " 'Usp33',\n",
       " 'Ust',\n",
       " 'Vapa',\n",
       " 'Vav3',\n",
       " 'Virma',\n",
       " 'Vmn2r18',\n",
       " 'Vps13b',\n",
       " 'Vwa3b',\n",
       " 'Vwc2l',\n",
       " 'Wdr72',\n",
       " 'Wif1',\n",
       " 'Zcchc7',\n",
       " 'Zfp142',\n",
       " 'Zfp563',\n",
       " 'Zfp600',\n",
       " 'Zfp629',\n",
       " 'Zfp773',\n",
       " 'Zfp954',\n",
       " 'Zfp978',\n",
       " 'Zfp979',\n",
       " 'Zfp982',\n",
       " 'Zfp985',\n",
       " 'Zfp987',\n",
       " 'Zfp988',\n",
       " 'Zfp989',\n",
       " 'Zfp991',\n",
       " 'Zfpm2',\n",
       " 'Zzz3']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniq_count = gene_tmp[\"marker_symbol\"].unique().shape[0]\n",
    "print(f\"unique genomic features: {uniq_count}\")\n",
    "\n",
    "candidate_genes = gene_tmp[\"marker_symbol\"].unique()\n",
    "[ x for x in sorted(candidate_genes) ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What about breaking up large CISs? It appears they are too large and span multiple genes\n",
    "\n",
    "# Or do now we check each individual gene that was found to further refine our search?\n",
    "\n",
    "# What's stopping me from testing literally every gene with +/-50 kb region of insertions with a ranksums test?\n",
    "# because we don't want this to be gene centric, however, it's not a bad way to validate the candidate gene list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rpl48-ps1 1 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3081</th>\n",
       "      <td>case</td>\n",
       "      <td>chr10</td>\n",
       "      <td>292</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>57474423</td>\n",
       "      <td>57474423</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type  chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "3081  case  chr10       292      1      0            4  57474423  57474423   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "3081      0                 1             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3544</th>\n",
       "      <td>control</td>\n",
       "      <td>chr10</td>\n",
       "      <td>119</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>57513025</td>\n",
       "      <td>57513025</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         type  chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "3544  control  chr10       119      1      0            1  57513025  57513025   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "3544      0                 1             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cpd 0 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4359</th>\n",
       "      <td>control</td>\n",
       "      <td>chr11</td>\n",
       "      <td>142</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>76722877</td>\n",
       "      <td>76722877</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         type  chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "4359  control  chr11       142      1      0            1  76722877  76722877   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "4359      0                 1             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ckb 0 0\n",
      "Ehhadh 1 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>case</td>\n",
       "      <td>chr16</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>1891</td>\n",
       "      <td>790</td>\n",
       "      <td>21600392</td>\n",
       "      <td>21614302</td>\n",
       "      <td>13910</td>\n",
       "      <td>62</td>\n",
       "      <td>602</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     type  chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "351  case  chr16         1     62   1891          790  21600392  21614302   \n",
       "\n",
       "     range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "351  13910                62           602                   32  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>410</th>\n",
       "      <td>control</td>\n",
       "      <td>chr16</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "      <td>325</td>\n",
       "      <td>89</td>\n",
       "      <td>21614194</td>\n",
       "      <td>21614302</td>\n",
       "      <td>108</td>\n",
       "      <td>26</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        type  chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "410  control  chr16         1     26    325           89  21614194  21614302   \n",
       "\n",
       "     range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "410    108                26             7                   14  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Son 0 0\n",
      "Macrod2 2 4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>725</th>\n",
       "      <td>case</td>\n",
       "      <td>chr2</td>\n",
       "      <td>83</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>141177054</td>\n",
       "      <td>141177560</td>\n",
       "      <td>506</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>929</th>\n",
       "      <td>case</td>\n",
       "      <td>chr2</td>\n",
       "      <td>287</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>141377758</td>\n",
       "      <td>141377758</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     type chrom  subgraph  nodes  edges  num_inserts    min_pos    max_pos  \\\n",
       "725  case  chr2        83      2      1            2  141177054  141177560   \n",
       "929  case  chr2       287      1      0            1  141377758  141377758   \n",
       "\n",
       "     range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "725    506                 2             0                    0  \n",
       "929      0                 1             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>963</th>\n",
       "      <td>control</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>34</td>\n",
       "      <td>140313463</td>\n",
       "      <td>140313466</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084</th>\n",
       "      <td>control</td>\n",
       "      <td>chr2</td>\n",
       "      <td>217</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>140190257</td>\n",
       "      <td>140190257</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1121</th>\n",
       "      <td>control</td>\n",
       "      <td>chr2</td>\n",
       "      <td>254</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>140580324</td>\n",
       "      <td>140580324</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1153</th>\n",
       "      <td>control</td>\n",
       "      <td>chr2</td>\n",
       "      <td>286</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>140753859</td>\n",
       "      <td>140753859</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         type chrom  subgraph  nodes  edges  num_inserts    min_pos  \\\n",
       "963   control  chr2        96      2      1           34  140313463   \n",
       "1084  control  chr2       217      1      0            1  140190257   \n",
       "1121  control  chr2       254      1      0            1  140580324   \n",
       "1153  control  chr2       286      1      0           15  140753859   \n",
       "\n",
       "        max_pos  range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "963   140313466      3                 2             0                    0  \n",
       "1084  140190257      0                 1             0                    0  \n",
       "1121  140580324      0                 1             0                    0  \n",
       "1153  140753859      0                 1             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sprr1b 1 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2218</th>\n",
       "      <td>case</td>\n",
       "      <td>chr3</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>438</td>\n",
       "      <td>92347697</td>\n",
       "      <td>92347714</td>\n",
       "      <td>17</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "2218  case  chr3         7      6     15          438  92347697  92347714   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "2218     17                 6             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2850</th>\n",
       "      <td>control</td>\n",
       "      <td>chr3</td>\n",
       "      <td>145</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>92347698</td>\n",
       "      <td>92347698</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         type chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "2850  control  chr3       145      1      0           41  92347698  92347698   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "2850      0                 1             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Eif3b 0 0\n",
      "Aak1 2 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5793</th>\n",
       "      <td>case</td>\n",
       "      <td>chr6</td>\n",
       "      <td>1</td>\n",
       "      <td>115</td>\n",
       "      <td>6555</td>\n",
       "      <td>10437</td>\n",
       "      <td>86910504</td>\n",
       "      <td>86911206</td>\n",
       "      <td>702</td>\n",
       "      <td>115</td>\n",
       "      <td>19</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5839</th>\n",
       "      <td>case</td>\n",
       "      <td>chr6</td>\n",
       "      <td>47</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>86858016</td>\n",
       "      <td>86858023</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "5793  case  chr6         1    115   6555        10437  86910504  86911206   \n",
       "5839  case  chr6        47      3      3           16  86858016  86858023   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "5793    702               115            19                   26  \n",
       "5839      7                 3             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7469</th>\n",
       "      <td>control</td>\n",
       "      <td>chr6</td>\n",
       "      <td>1</td>\n",
       "      <td>77</td>\n",
       "      <td>2926</td>\n",
       "      <td>845</td>\n",
       "      <td>86886566</td>\n",
       "      <td>86910899</td>\n",
       "      <td>24333</td>\n",
       "      <td>77</td>\n",
       "      <td>881</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7679</th>\n",
       "      <td>control</td>\n",
       "      <td>chr6</td>\n",
       "      <td>211</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>86976664</td>\n",
       "      <td>86976664</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         type chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "7469  control  chr6         1     77   2926          845  86886566  86910899   \n",
       "7679  control  chr6       211      1      0           55  86976664  86976664   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "7469  24333                77           881                   19  \n",
       "7679      0                 1             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rnf214 1 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3108</th>\n",
       "      <td>case</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>36</td>\n",
       "      <td>774</td>\n",
       "      <td>45852161</td>\n",
       "      <td>45852171</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type chrom  subgraph  nodes  edges  num_inserts   min_pos   max_pos  \\\n",
       "3108  case  chr9         5      9     36          774  45852161  45852171   \n",
       "\n",
       "      range  num_insert_sites  num_ta_sites  num_ta_insert_sites  \n",
       "3108     10                 9             0                    0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# TODO: use list of Laura genes and see why that genomic region did not show up\n",
    "top10genes = [\"Aak1\", \"Ehhadh\", \"Macrod2\", \"Ckb\", \"Rnf214\", \"Sprr1b\", \"Cpd\", \"Rpl48-ps1\", \"Son\", \"Eif3b\"]\n",
    "top10annot = annot_df[annot_df[\"Marker Symbol\"].isin(top10genes)]\n",
    "for row in top10annot.itertuples():\n",
    "    chrom = f\"chr{row.Chr}\"\n",
    "    case_chrom_df = case_df[case_df[\"chrom\"] == chrom]\n",
    "    control_chrom_df = control_df[control_df[\"chrom\"] == chrom]\n",
    "    res1 = int(row._4 - 50000) <= case_chrom_df[\"max_pos\"]\n",
    "    res2 = int(row._5 + 50000) >= case_chrom_df[\"min_pos\"]\n",
    "    res3 = int(row._4 - 50000) <= control_chrom_df[\"max_pos\"]\n",
    "    res4 = int(row._5 + 50000) >= control_chrom_df[\"min_pos\"]\n",
    "\n",
    "    print(row._7, (res1 & res2).sum(), (res3 & res4).sum())\n",
    "    if (res1 & res2).sum() != 0:\n",
    "        display(case_chrom_df.loc[res1 & res2])\n",
    "    if (res3 & res4).sum() != 0:\n",
    "        display(control_chrom_df.loc[res3 & res4])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: check for each gene with an extended proomoter the case vs controls insertions\n",
    "# check with ranksums test as well as binomial test\n",
    "# What does the candidate gene list look like now?\n",
    "\n",
    "\n",
    "# TODO: I need to go through my code again. Go step by step for each function in another notebook\n",
    "# double check this code and cis_networks.py\n",
    "\n",
    "\n",
    "# TODO: rerun without mapq thresholding in preprocess_reads and preprocess_insertions\n",
    "\n",
    "\n",
    "# TODO: should we be using normalized read counts? idk, what stage would this occur at\n",
    "# the variation would come from read depth, so we would have to normalize based on this BEFORE preprocess_reads.py\n",
    "# What is done with RNAseq? How to normalize on read depth while still keeping count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# volcano plot of p-values and LFC for each CIS if wanted\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output list of pCIS (now can be called CIS) that are sig.\n",
    "# per CIS, output list of TAs that are sig.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# take identified candidate CIS/genes and check for co-occurrence with all other CIS/genes\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "netcis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
