{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Why do IRL and IRR files have similar, if not the same, insertions?\n",
    "\n",
    "Could be because the transposon and adapter tags were incorrect from the IAS pipeline, and on top of that they were incorrectly used in cutadapt (not using reverse complement). I have output from bowtie2 that summarizes the mapping results and will need to dynamically parse the output file for mapping qualities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "bowtie_dir = Path(\"/project/cs-myers/MathewF/projects/Laura-SB-Analysis/NetCIS/toy-data/2020_SB-bowtie/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>tumor_type</th>\n",
       "      <th>tumor_library</th>\n",
       "      <th>total-reads</th>\n",
       "      <th>paired-reads</th>\n",
       "      <th>concordant-reads-0</th>\n",
       "      <th>concordant-reads-1</th>\n",
       "      <th>concordant-reads-&gt;1</th>\n",
       "      <th>discordant-reads-1</th>\n",
       "      <th>aligned-reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B16</td>\n",
       "      <td>1_1</td>\n",
       "      <td>RT</td>\n",
       "      <td>IRR</td>\n",
       "      <td>1529</td>\n",
       "      <td>1523</td>\n",
       "      <td>1104</td>\n",
       "      <td>394</td>\n",
       "      <td>25</td>\n",
       "      <td>3</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B16</td>\n",
       "      <td>1_1</td>\n",
       "      <td>S</td>\n",
       "      <td>IRR</td>\n",
       "      <td>1810</td>\n",
       "      <td>1795</td>\n",
       "      <td>1625</td>\n",
       "      <td>153</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B16</td>\n",
       "      <td>1_1</td>\n",
       "      <td>RT</td>\n",
       "      <td>IRL</td>\n",
       "      <td>1733</td>\n",
       "      <td>1733</td>\n",
       "      <td>1706</td>\n",
       "      <td>4</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B16</td>\n",
       "      <td>1_1</td>\n",
       "      <td>S</td>\n",
       "      <td>IRL</td>\n",
       "      <td>3094</td>\n",
       "      <td>3094</td>\n",
       "      <td>3085</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cell_type sample_id tumor_type tumor_library  total-reads  paired-reads  \\\n",
       "0       B16       1_1         RT           IRR         1529          1523   \n",
       "1       B16       1_1          S           IRR         1810          1795   \n",
       "2       B16       1_1         RT           IRL         1733          1733   \n",
       "3       B16       1_1          S           IRL         3094          3094   \n",
       "\n",
       "   concordant-reads-0  concordant-reads-1  concordant-reads->1  \\\n",
       "0                1104                 394                   25   \n",
       "1                1625                 153                   17   \n",
       "2                1706                   4                   23   \n",
       "3                3085                   3                    6   \n",
       "\n",
       "   discordant-reads-1  aligned-reads  \n",
       "0                   3            422  \n",
       "1                   1            171  \n",
       "2                   0             27  \n",
       "3                   0              9  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "list_df = []\n",
    "for file in bowtie_dir.iterdir():\n",
    "    cell_type, sample_id, tumor_tmp = file.name.split(\"-\")\n",
    "    tumor_type, lib_tmp = tumor_tmp.split(\"_\")\n",
    "    library_type = lib_tmp.split(\".\")[0]\n",
    "    tmp_dict = {\"cell_type\": cell_type, \"sample_id\": sample_id, \"tumor_type\": tumor_type, \"tumor_library\": library_type}\n",
    "    with open(file) as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.find(\"reads; of these:\") != -1:\n",
    "                tmp_dict[\"total-reads\"] = [int(line.split(\" \")[0])]\n",
    "            elif line.find(\"were paired; of these:\") != -1:\n",
    "                tmp_dict[\"paired-reads\"] = [int(line.split(\" \")[0])]\n",
    "            elif line.find(\"aligned concordantly 0 times\") != -1:\n",
    "                tmp_dict[\"concordant-reads-0\"] = [int(line.split(\" \")[0])]\n",
    "            elif line.find(\"aligned concordantly exactly 1 time\") != -1:\n",
    "                tmp_dict[\"concordant-reads-1\"] = [int(line.split(\" \")[0])]\n",
    "            elif line.find(\"aligned concordantly >1 times\") != -1:\n",
    "                tmp_dict[\"concordant-reads->1\"] = [int(line.split(\" \")[0])]\n",
    "            elif line.find(\"aligned discordantly 1 time\") != -1:\n",
    "                tmp_dict[\"discordant-reads-1\"] = [int(line.split(\" \")[0])]\n",
    "            else:\n",
    "                pass\n",
    "    tmp_df = pd.DataFrame.from_dict(tmp_dict)\n",
    "    list_df.append(tmp_df)\n",
    "bowtie_df = pd.concat(list_df, ignore_index=True)\n",
    "bowtie_df[\"aligned-reads\"] = bowtie_df[\"concordant-reads-1\"] + bowtie_df[\"concordant-reads->1\"] + bowtie_df[\"discordant-reads-1\"]\n",
    "display(bowtie_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# can also check stats of the bam file with:\n",
    "# samtools flagstat <file.bam>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "netcis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
