{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "from pathlib import Path\n",
    "from multiprocessing import Pool\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn.objects as so\n",
    "from seaborn import axes_style\n",
    "import networkx as nx\n",
    "from scipy.stats import binomtest, ranksums, mannwhitneyu, wilcoxon, skewtest, kurtosistest\n",
    "from tqdm import tqdm\n",
    "\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "dir1 = os.path.dirname(os.path.abspath(''))\n",
    "if not dir1 in sys.path: sys.path.append(dir1)\n",
    "from netcis import network_analysis as na\n",
    "from importlib import reload\n",
    "# reload(na)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "refdata = Path(\"/project/cs-myers/MathewF/projects/Laura-SB-Analysis/2023-SB-screen/ref_data/GRCm39\")\n",
    "\n",
    "args = {\n",
    "    \"output_prefix\": \"/project/cs-myers/MathewF/projects/Laura-SB-Analysis/2023-SB-screen/output/GRCm39/results\", \n",
    "    \"ta_dir\": refdata / \"ta_files\",\n",
    "    \"gene_annot\": refdata / \"MRK_List2.rpt\",\n",
    "    \"ta_error\": 5,\n",
    "    \"pval_threshold\": 0.05,\n",
    "    \"verbose\": 1,\n",
    "    \"case\": \"CAR\",  # CAR ACF LT RT\n",
    "    \"control\": \"NoCAR\",  # NoCAR SCF S S\n",
    "    \"npara\": 21,\n",
    "}\n",
    "\n",
    "args[\"graph_dir\"] = Path(args[\"output_prefix\"] + \"-graphs/\")\n",
    "\n",
    "output = Path(args[\"output_prefix\"] + \"-analysis\")\n",
    "output.mkdir(exist_ok=True)\n",
    "\n",
    "ta_dir = args[\"ta_dir\"]\n",
    "gene_annot = args[\"gene_annot\"]\n",
    "ta_error = args[\"ta_error\"]\n",
    "pval_threshold = args[\"pval_threshold\"]\n",
    "verbose = args[\"verbose\"]\n",
    "case = args[\"case\"]\n",
    "control = args[\"control\"]\n",
    "\n",
    "output_res = output / f\"{case}-{control}\"\n",
    "output_res.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chr1', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chrX', 'chrY']\n",
      "21\n"
     ]
    }
   ],
   "source": [
    "annot_df = pd.read_csv(gene_annot, sep=\"\\t\")\n",
    "annot_df = annot_df[pd.notna(annot_df[\"genome coordinate start\"])].drop(\"Status\", axis=1)\n",
    "annot_df[\"chrom\"] = annot_df[\"Chr\"].apply(lambda x: f\"chr{x}\")\n",
    "annot_df = annot_df.sort_values([\"chrom\"]).reset_index(drop=True)\n",
    "# TODO: what about the strand in annot_df?\n",
    "\n",
    "bed_files = {file.name.split(\".\")[0]: file for file in args[\"ta_dir\"].iterdir()}\n",
    "\n",
    "chroms = sorted([ chrom.name for chrom in (args[\"graph_dir\"] / case).iterdir() ])\n",
    "print(chroms)\n",
    "print(len(chroms))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrY\tsig. genomic features: 6/2258\n",
      "chr19\tsig. genomic features: 185/19111\n",
      "chr18\tsig. genomic features: 228/19513\n",
      "chr17\tsig. genomic features: 423/27642\n",
      "chr16\tsig. genomic features: 258/21700\n",
      "chr14\tsig. genomic features: 300/25888\n",
      "chr15\tsig. genomic features: 373/25472\n",
      "chr12\tsig. genomic features: 409/26317\n",
      "chr10\tsig. genomic features: 324/31354\n",
      "chr11\tsig. genomic features: 509/43017\n",
      "chr13\tsig. genomic features: 497/27823\n",
      "chr8\tsig. genomic features: 456/31843\n",
      "chrX\tsig. genomic features: 275/17094\n",
      "chr3\tsig. genomic features: 390/32790\n",
      "chr7\tsig. genomic features: 475/41167\n",
      "chr5\tsig. genomic features: 658/41312\n",
      "chr6\tsig. genomic features: 532/36110\n",
      "chr2\tsig. genomic features: 847/48061\n",
      "chr9\tsig. genomic features: 1295/35050\n",
      "chr1\tsig. genomic features: 1064/42676\n",
      "chr4\tsig. genomic features: 2202/38706\n"
     ]
    }
   ],
   "source": [
    "reload(na)\n",
    "# get chromosome subsets for annotation file and TA bed file\n",
    "# iter_args = tqdm([ (chrom, annot_df[annot_df[\"chrom\"] == chrom], bed_files[chrom], args) for chrom in chroms ])\n",
    "iter_args = [ (chrom, annot_df[annot_df[\"chrom\"] == chrom], bed_files[chrom], args) for chrom in chroms ]\n",
    "with Pool(args[\"npara\"]) as p:\n",
    "    res_dict_list = [ x for x in p.imap_unordered(na.chrom_analysis, iter_args) ]\n",
    "# res_dict_list = na.chrom_analysis(iter_args[-2])\n",
    "\n",
    "\n",
    "# join chromosomes results together\n",
    "ta_list = []\n",
    "overall_list = []\n",
    "sig_list = []\n",
    "genomic_features_list = []\n",
    "graphs_stats = []\n",
    "for res_dict in res_dict_list:\n",
    "    ta_list.append(res_dict[\"ta\"])\n",
    "    overall_list.append(res_dict[\"overall\"])\n",
    "    sig_list.append(res_dict[\"sig\"])\n",
    "    genomic_features_list.append(res_dict[\"genomic_features\"])\n",
    "    graphs_stats.append(res_dict[\"graph_stats\"])\n",
    "\n",
    "TA_df = pd.concat(ta_list, ignore_index=True)\n",
    "overall_df = pd.concat(overall_list, ignore_index=True)\n",
    "sig_df = pd.concat(sig_list, ignore_index=True)\n",
    "genomic_features_df = pd.concat(genomic_features_list, ignore_index=True)\n",
    "graph_stats_df = pd.concat(graphs_stats, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "381 380\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_index</th>\n",
       "      <th>reference_index</th>\n",
       "      <th>target_pos_min</th>\n",
       "      <th>target_pos_max</th>\n",
       "      <th>reference_pos_min</th>\n",
       "      <th>reference_pos_max</th>\n",
       "      <th>mannwhitneyu</th>\n",
       "      <th>ranksums</th>\n",
       "      <th>binomial</th>\n",
       "      <th>target_num_samples</th>\n",
       "      <th>reference_num_samples</th>\n",
       "      <th>total_IS</th>\n",
       "      <th>sig_IS</th>\n",
       "      <th>target_IS_count</th>\n",
       "      <th>reference_IS_count</th>\n",
       "      <th>sig_ratio</th>\n",
       "      <th>target</th>\n",
       "      <th>reference</th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>103846357</td>\n",
       "      <td>103846381</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.021071</td>\n",
       "      <td>0.020921</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "      <td>None</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>169</td>\n",
       "      <td>0</td>\n",
       "      <td>0.75</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>CAR</td>\n",
       "      <td>chr5</td>\n",
       "      <td>1700016H13Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>138825098</td>\n",
       "      <td>138825098</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2310009B15Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27307799</td>\n",
       "      <td>27307799</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr7</td>\n",
       "      <td>2310022A10Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50676643</td>\n",
       "      <td>50676646</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.059346</td>\n",
       "      <td>0.049535</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>84</td>\n",
       "      <td>0</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr9</td>\n",
       "      <td>2310030G06Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>145654994</td>\n",
       "      <td>145654994</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>CAR</td>\n",
       "      <td>chr3</td>\n",
       "      <td>2410004B18Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>116</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147053363</td>\n",
       "      <td>147053363</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "      <td>Zfp989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>146675191</td>\n",
       "      <td>146675191</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>0.000977</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "      <td>Zfp993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25507102</td>\n",
       "      <td>25507102</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>0.000977</td>\n",
       "      <td>8</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr14</td>\n",
       "      <td>Zmiz1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>126941561</td>\n",
       "      <td>126941597</td>\n",
       "      <td>126941571</td>\n",
       "      <td>126941601</td>\n",
       "      <td>0.325364</td>\n",
       "      <td>0.306685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>48</td>\n",
       "      <td>142</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "      <td>Zmym1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>126941561</td>\n",
       "      <td>126941597</td>\n",
       "      <td>126941571</td>\n",
       "      <td>126941601</td>\n",
       "      <td>0.325364</td>\n",
       "      <td>0.306685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>48</td>\n",
       "      <td>142</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "      <td>Zmym6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>381 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    target_index reference_index target_pos_min target_pos_max  \\\n",
       "0              2             NaN      103846357      103846381   \n",
       "1            125             NaN      138825098      138825098   \n",
       "2             38             NaN       27307799       27307799   \n",
       "3              6             NaN       50676643       50676646   \n",
       "4             35             NaN      145654994      145654994   \n",
       "..           ...             ...            ...            ...   \n",
       "376          116             NaN      147053363      147053363   \n",
       "377          139             NaN      146675191      146675191   \n",
       "378            6             NaN       25507102       25507102   \n",
       "379            9             0.0      126941561      126941597   \n",
       "380            9             0.0      126941561      126941597   \n",
       "\n",
       "    reference_pos_min reference_pos_max mannwhitneyu  ranksums  binomial  \\\n",
       "0                None              None     0.021071  0.020921       0.0   \n",
       "1                None              None          1.0  0.317311   0.03125   \n",
       "2                None              None          1.0  0.317311  0.000122   \n",
       "3                None              None     0.059346  0.049535       0.0   \n",
       "4                None              None          1.0  0.317311  0.000015   \n",
       "..                ...               ...          ...       ...       ...   \n",
       "376              None              None          1.0  0.317311   0.03125   \n",
       "377              None              None          1.0  0.317311  0.000977   \n",
       "378              None              None          1.0  0.317311  0.000977   \n",
       "379         126941571         126941601     0.325364  0.306685       0.0   \n",
       "380         126941571         126941601     0.325364  0.306685       0.0   \n",
       "\n",
       "    target_num_samples reference_num_samples total_IS sig_IS target_IS_count  \\\n",
       "0                    3                  None        4      3             169   \n",
       "1                    1                  None        1      1               6   \n",
       "2                    2                  None        1      1              14   \n",
       "3                    2                  None        3      1              84   \n",
       "4                    1                  None        1      1              17   \n",
       "..                 ...                   ...      ...    ...             ...   \n",
       "376                  1                  None        1      1               6   \n",
       "377                  1                  None        1      1              11   \n",
       "378                  8                  None        1      1              11   \n",
       "379                  4                     6        7      3              48   \n",
       "380                  4                     6        7      3              48   \n",
       "\n",
       "    reference_IS_count sig_ratio target reference  chrom           gene  \n",
       "0                    0      0.75  NoCAR       CAR   chr5  1700016H13Rik  \n",
       "1                    0       1.0    CAR     NoCAR   chr1  2310009B15Rik  \n",
       "2                    0       1.0    CAR     NoCAR   chr7  2310022A10Rik  \n",
       "3                    0  0.333333    CAR     NoCAR   chr9  2310030G06Rik  \n",
       "4                    0       1.0  NoCAR       CAR   chr3  2410004B18Rik  \n",
       "..                 ...       ...    ...       ...    ...            ...  \n",
       "376                  0       1.0    CAR     NoCAR   chr4         Zfp989  \n",
       "377                  0       1.0    CAR     NoCAR   chr4         Zfp993  \n",
       "378                  0       1.0    CAR     NoCAR  chr14          Zmiz1  \n",
       "379                142  0.428571    CAR     NoCAR   chr4          Zmym1  \n",
       "380                142  0.428571    CAR     NoCAR   chr4          Zmym6  \n",
       "\n",
       "[381 rows x 20 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CAR</th>\n",
       "      <th>NoCAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pCIS_count</th>\n",
       "      <td>1220</td>\n",
       "      <td>1128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insertion_site_count</th>\n",
       "      <td>1663</td>\n",
       "      <td>1529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads_count</th>\n",
       "      <td>22334</td>\n",
       "      <td>15628</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        CAR  NoCAR\n",
       "pCIS_count             1220   1128\n",
       "insertion_site_count   1663   1529\n",
       "reads_count           22334  15628"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove duplicated entries\n",
    "def check_dups(gene_df, case, control):\n",
    "    if len(gene_df) == 1:\n",
    "        return gene_df\n",
    "    \n",
    "    case_tmp = gene_df[gene_df[\"target\"] == case]\n",
    "    cont_tmp = gene_df[gene_df[\"target\"] == control]\n",
    "\n",
    "    return_list = []\n",
    "    remove_list = []\n",
    "    for _, row in case_tmp.iterrows():\n",
    "        t1 = cont_tmp[\"reference_index\"].values == row[\"target_index\"]\n",
    "        t2 = cont_tmp[\"reference_IS_count\"].values == row[\"target_IS_count\"]\n",
    "        if any(t1) and any(t2) and np.array_equal(t1, t2):\n",
    "            return_list.append(row.to_frame().T)\n",
    "            remove_list.append(cont_tmp[t1].index)\n",
    "            \n",
    "    if remove_list:\n",
    "        remove_list = remove_list[0] if len(remove_list) == 1 else [ x[0] for x in remove_list]\n",
    "        new_cont_tmp = cont_tmp.drop(index=remove_list)\n",
    "        for _, row in new_cont_tmp.iterrows():\n",
    "            t1 = case_tmp[\"target_index\"].values == row[\"reference_index\"]\n",
    "            t2 = case_tmp[\"target_IS_count\"].values == row[\"reference_IS_count\"]\n",
    "            if any(t1) and any(t2) and np.array_equal(t1, t2):\n",
    "                return_list.append(row.to_frame().T)\n",
    "            \n",
    "    if not return_list:\n",
    "        return None\n",
    "    else:\n",
    "        return pd.concat(return_list)\n",
    "            \n",
    "# get candidate gene list \n",
    "genes_tmp = genomic_features_df[(genomic_features_df[\"marker_type\"] == \"Gene\") & (genomic_features_df[\"marker_feature_type\"] == \"protein coding gene\")]\n",
    "# genes_only = genes_tmp.groupby([\"type_index\"]).agg(list)[\"marker_symbol\"].reset_index()\n",
    "many_genes = pd.DataFrame({\"marker_symbol\": [ x for x in sorted(genes_tmp[\"marker_symbol\"].unique()) ]})\n",
    "\n",
    "new_genes = []\n",
    "for i, gene in enumerate(many_genes[\"marker_symbol\"].values):\n",
    "    tmp_annot = genomic_features_df[genomic_features_df[\"marker_symbol\"] == gene]\n",
    "    # display(tmp_annot)\n",
    "    \n",
    "    tmp_sig = sig_df[sig_df[\"target_index\"].isin(tmp_annot[\"type_index\"]) & sig_df[\"chrom\"].isin(tmp_annot[\"chrom\"])].sort_values([\"target_index\"])\n",
    "    tmp_sig[\"gene\"] = gene\n",
    "    tmp_gene = check_dups(tmp_sig, case, control)\n",
    "    if tmp_gene is not None:\n",
    "        new_genes.append(tmp_gene)\n",
    "        # display(tmp_gene)\n",
    "    \n",
    "        # ta_list = []\n",
    "        # for ta_subset in tmp_gene.itertuples():\n",
    "        #     tmp = TA_df[TA_df[\"target_index\"].isin([ta_subset.target_index]) & TA_df[\"chrom\"].isin([ta_subset.chrom]) & TA_df[\"target\"].isin([ta_subset.target])]\n",
    "        #     ta_list.append(tmp)\n",
    "        # tmp_ta = pd.concat(ta_list, ignore_index=True).sort_values([\"target_index\", \"reference_index\", \"target\", \"pos\"])\n",
    "        # print(pearsonr(tmp_ta[\"target_count\"].values, tmp_ta[\"reference_count\"].values))\n",
    "        # display(tmp_ta)\n",
    "    \n",
    "    # stats_list = []\n",
    "    # for ta_subset in tmp_annot.itertuples():\n",
    "    #     tmp = graph_stats_df[graph_stats_df[\"subgraph\"].isin([ta_subset.type_index]) & graph_stats_df[\"chrom\"].isin([ta_subset.chrom]) & graph_stats_df[\"type\"].isin([ta_subset.type_name])]\n",
    "    #     stats_list.append(tmp)\n",
    "    # tmp_g_stats = pd.concat(stats_list, ignore_index=True).sort_values([\"subgraph\"])\n",
    "    # # display(tmp_g_stats)\n",
    "\n",
    "    \n",
    "\n",
    "candidate_genes = pd.concat(new_genes, ignore_index=True)\n",
    "a = candidate_genes[\"target_IS_count\"][candidate_genes[\"target\"] == case].rename(case)\n",
    "b = candidate_genes[\"reference_IS_count\"][candidate_genes[\"reference\"] == case].rename(case)\n",
    "case_read_counts = pd.concat([a, b]).sort_index()\n",
    "c = candidate_genes[\"target_IS_count\"][candidate_genes[\"target\"] == control].rename(control)\n",
    "d = candidate_genes[\"reference_IS_count\"][candidate_genes[\"reference\"] == control].rename(control)\n",
    "control_read_counts = pd.concat([c, d]).sort_index()\n",
    "\n",
    "output_genes = candidate_genes.groupby([\"gene\"])[[\"mannwhitneyu\", \"ranksums\", \"binomial\"]].mean().reset_index()\n",
    "\n",
    "print(len(candidate_genes), len(output_genes))\n",
    "display(candidate_genes)\n",
    "output_genes.columns = [\"gene\", \"avg-mannwhitneyu\", \"avg-ranksums\",  \"avg-binomial\"]\n",
    "output_genes[case] = case_read_counts\n",
    "output_genes[control] = control_read_counts\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# save data\n",
    "TA_df.to_csv(output_res / \"TA.tsv\", sep=\"\\t\", index=False)\n",
    "overall_df.to_csv(output_res / \"overall.tsv\", sep=\"\\t\", index=False)\n",
    "sig_df.to_csv(output_res / \"sig.tsv\", sep=\"\\t\", index=False)\n",
    "genomic_features_df.to_csv(output_res / \"genomic_features.tsv\", sep=\"\\t\", index=False)\n",
    "graph_stats_df.to_csv(output_res / \"graph_stats.tsv\", sep=\"\\t\", index=False)\n",
    "candidate_genes.to_csv(output_res / \"candidate_genes.tsv\", sep=\"\\t\", index=False)\n",
    "output_genes.to_csv(output_res / \"cleaned_output_genes.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "simple_summary = {}\n",
    "for treatment in graph_stats_df[\"type\"].unique():\n",
    "    treatment_df = graph_stats_df[graph_stats_df[\"type\"] == treatment]\n",
    "    simple_summary[treatment] = {\"pCIS_count\": len(treatment_df),\n",
    "                                 \"insertion_site_count\": treatment_df[\"nodes\"].sum(),\n",
    "                                 \"reads_count\": treatment_df[\"num_inserts\"].sum(),\n",
    "                                 }\n",
    "df = pd.DataFrame(simple_summary)\n",
    "df.to_csv(args['graph_dir'].parent / \"simple_summary.csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>subgraph</th>\n",
       "      <th>nodes</th>\n",
       "      <th>edges</th>\n",
       "      <th>num_inserts</th>\n",
       "      <th>min_pos</th>\n",
       "      <th>max_pos</th>\n",
       "      <th>range</th>\n",
       "      <th>sample_IDs</th>\n",
       "      <th>num_unique_samples</th>\n",
       "      <th>num_insert_sites</th>\n",
       "      <th>num_ta_sites</th>\n",
       "      <th>num_ta_insert_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAR</td>\n",
       "      <td>chrY</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>38017098</td>\n",
       "      <td>38017098</td>\n",
       "      <td>0</td>\n",
       "      <td>[4_4_8]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAR</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>12772988</td>\n",
       "      <td>12772988</td>\n",
       "      <td>0</td>\n",
       "      <td>[4_4_14]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAR</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>17371977</td>\n",
       "      <td>17371977</td>\n",
       "      <td>0</td>\n",
       "      <td>[4_4_14]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAR</td>\n",
       "      <td>chrY</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>61879714</td>\n",
       "      <td>61879714</td>\n",
       "      <td>0</td>\n",
       "      <td>[4_4_14]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAR</td>\n",
       "      <td>chrY</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>62423308</td>\n",
       "      <td>62423308</td>\n",
       "      <td>0</td>\n",
       "      <td>[4_4_14]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2343</th>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "      <td>182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>94580864</td>\n",
       "      <td>94580864</td>\n",
       "      <td>0</td>\n",
       "      <td>[4_4_12]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2344</th>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "      <td>183</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>101363952</td>\n",
       "      <td>101363952</td>\n",
       "      <td>0</td>\n",
       "      <td>[4_4_12]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2345</th>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "      <td>184</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>122337858</td>\n",
       "      <td>122337858</td>\n",
       "      <td>0</td>\n",
       "      <td>[4_4_12]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2346</th>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "      <td>185</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>41142724</td>\n",
       "      <td>41142724</td>\n",
       "      <td>0</td>\n",
       "      <td>[3_20_17]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2347</th>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "      <td>186</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>51126838</td>\n",
       "      <td>51126838</td>\n",
       "      <td>0</td>\n",
       "      <td>[3_20_17]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2348 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       type chrom  subgraph  nodes  edges  num_inserts    min_pos    max_pos  \\\n",
       "0       CAR  chrY         0      1      0            1   38017098   38017098   \n",
       "1       CAR  chrY         1      1      0            1   12772988   12772988   \n",
       "2       CAR  chrY         2      1      0            1   17371977   17371977   \n",
       "3       CAR  chrY         3      1      0            1   61879714   61879714   \n",
       "4       CAR  chrY         4      1      0            1   62423308   62423308   \n",
       "...     ...   ...       ...    ...    ...          ...        ...        ...   \n",
       "2343  NoCAR  chr4       182      1      0            6   94580864   94580864   \n",
       "2344  NoCAR  chr4       183      1      0            2  101363952  101363952   \n",
       "2345  NoCAR  chr4       184      1      0            5  122337858  122337858   \n",
       "2346  NoCAR  chr4       185      1      0            1   41142724   41142724   \n",
       "2347  NoCAR  chr4       186      1      0            1   51126838   51126838   \n",
       "\n",
       "      range sample_IDs  num_unique_samples  num_insert_sites  num_ta_sites  \\\n",
       "0         0    [4_4_8]                   1                 1             0   \n",
       "1         0   [4_4_14]                   1                 1             0   \n",
       "2         0   [4_4_14]                   1                 1             0   \n",
       "3         0   [4_4_14]                   1                 1             0   \n",
       "4         0   [4_4_14]                   1                 1             0   \n",
       "...     ...        ...                 ...               ...           ...   \n",
       "2343      0   [4_4_12]                   1                 1             0   \n",
       "2344      0   [4_4_12]                   1                 1             0   \n",
       "2345      0   [4_4_12]                   1                 1             0   \n",
       "2346      0  [3_20_17]                   1                 1             0   \n",
       "2347      0  [3_20_17]                   1                 1             0   \n",
       "\n",
       "      num_ta_insert_sites  \n",
       "0                       0  \n",
       "1                       0  \n",
       "2                       0  \n",
       "3                       0  \n",
       "4                       0  \n",
       "...                   ...  \n",
       "2343                    0  \n",
       "2344                    0  \n",
       "2345                    0  \n",
       "2346                    0  \n",
       "2347                    0  \n",
       "\n",
       "[2348 rows x 14 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_stats_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAR\n",
      "NoCAR\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>treatment</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAR</td>\n",
       "      <td>1220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NoCAR</td>\n",
       "      <td>1128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  treatment  count\n",
       "0       CAR   1220\n",
       "1     NoCAR   1128"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pCIS_counts = {}\n",
    "for treatment in graph_stats_df[\"type\"].unique():\n",
    "    total = len(graph_stats_df[graph_stats_df[\"type\"] == treatment])\n",
    "    pCIS_counts[treatment] = total\n",
    "treatment, count = zip(*pCIS_counts.items())\n",
    "df = pd.DataFrame({\"treatment\": treatment, \"count\": count})\n",
    "df.to_csv(args['graph_dir'].parent / \"pCIS_counts.csv\", index=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_index</th>\n",
       "      <th>reference_index</th>\n",
       "      <th>target_pos_min</th>\n",
       "      <th>target_pos_max</th>\n",
       "      <th>reference_pos_min</th>\n",
       "      <th>reference_pos_max</th>\n",
       "      <th>mannwhitneyu</th>\n",
       "      <th>ranksums</th>\n",
       "      <th>binomial</th>\n",
       "      <th>target_num_samples</th>\n",
       "      <th>reference_num_samples</th>\n",
       "      <th>total_IS</th>\n",
       "      <th>sig_IS</th>\n",
       "      <th>target_IS_count</th>\n",
       "      <th>reference_IS_count</th>\n",
       "      <th>sig_ratio</th>\n",
       "      <th>target</th>\n",
       "      <th>reference</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>54037701</td>\n",
       "      <td>54037707</td>\n",
       "      <td>54037705</td>\n",
       "      <td>54037705</td>\n",
       "      <td>0.026519</td>\n",
       "      <td>0.020921</td>\n",
       "      <td>0.00</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2518</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>72</td>\n",
       "      <td>15.0</td>\n",
       "      <td>79335535</td>\n",
       "      <td>79335535</td>\n",
       "      <td>79301992</td>\n",
       "      <td>79335535</td>\n",
       "      <td>0.414216</td>\n",
       "      <td>0.245278</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1254</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>CAR</td>\n",
       "      <td>chr9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>556</th>\n",
       "      <td>130</td>\n",
       "      <td>7.0</td>\n",
       "      <td>54037705</td>\n",
       "      <td>54037705</td>\n",
       "      <td>54037701</td>\n",
       "      <td>54037707</td>\n",
       "      <td>0.026519</td>\n",
       "      <td>0.020921</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2518</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>CAR</td>\n",
       "      <td>chr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>3</td>\n",
       "      <td>46.0</td>\n",
       "      <td>42652132</td>\n",
       "      <td>42652134</td>\n",
       "      <td>42652132</td>\n",
       "      <td>42652132</td>\n",
       "      <td>0.184039</td>\n",
       "      <td>0.126630</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1597</td>\n",
       "      <td>5</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>CAR</td>\n",
       "      <td>chr6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>15</td>\n",
       "      <td>0.0</td>\n",
       "      <td>79301992</td>\n",
       "      <td>79335535</td>\n",
       "      <td>79335535</td>\n",
       "      <td>79335535</td>\n",
       "      <td>0.414216</td>\n",
       "      <td>0.245278</td>\n",
       "      <td>0.00</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1254</td>\n",
       "      <td>1</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59015917</td>\n",
       "      <td>59037723</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.046854</td>\n",
       "      <td>0.049535</td>\n",
       "      <td>0.25</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>CAR</td>\n",
       "      <td>chr14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3051513</td>\n",
       "      <td>3053483</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.046854</td>\n",
       "      <td>0.049535</td>\n",
       "      <td>0.25</td>\n",
       "      <td>3</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>CAR</td>\n",
       "      <td>chr14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509</th>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>104343449</td>\n",
       "      <td>104376596</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.046854</td>\n",
       "      <td>0.049535</td>\n",
       "      <td>0.25</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>CAR</td>\n",
       "      <td>chr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87365491</td>\n",
       "      <td>87399417</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.046854</td>\n",
       "      <td>0.049535</td>\n",
       "      <td>0.25</td>\n",
       "      <td>3</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>CAR</td>\n",
       "      <td>chr9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>52520386</td>\n",
       "      <td>52520390</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.046854</td>\n",
       "      <td>0.049535</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>CAR</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>570 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     target_index  reference_index  target_pos_min  target_pos_max  \\\n",
       "430             7              0.0        54037701        54037707   \n",
       "360            72             15.0        79335535        79335535   \n",
       "556           130              7.0        54037705        54037705   \n",
       "260             3             46.0        42652132        42652134   \n",
       "312            15              0.0        79301992        79335535   \n",
       "..            ...              ...             ...             ...   \n",
       "57              2              NaN        59015917        59037723   \n",
       "56              1              NaN         3051513         3053483   \n",
       "509            21              NaN       104343449       104376596   \n",
       "342             8              NaN        87365491        87399417   \n",
       "45              2              NaN        52520386        52520390   \n",
       "\n",
       "    reference_pos_min reference_pos_max  mannwhitneyu  ranksums  binomial  \\\n",
       "430          54037705          54037705      0.026519  0.020921      0.00   \n",
       "360          79301992          79335535      0.414216  0.245278      0.00   \n",
       "556          54037701          54037707      0.026519  0.020921      0.00   \n",
       "260          42652132          42652132      0.184039  0.126630      0.00   \n",
       "312          79335535          79335535      0.414216  0.245278      0.00   \n",
       "..                ...               ...           ...       ...       ...   \n",
       "57               None              None      0.046854  0.049535      0.25   \n",
       "56               None              None      0.046854  0.049535      0.25   \n",
       "509              None              None      0.046854  0.049535      0.25   \n",
       "342              None              None      0.046854  0.049535      0.25   \n",
       "45               None              None      0.046854  0.049535      0.25   \n",
       "\n",
       "     target_num_samples reference_num_samples  total_IS  sig_IS  \\\n",
       "430                  12                     4         4       4   \n",
       "360                   1                     5         2       1   \n",
       "556                   4                    12         4       4   \n",
       "260                   7                     3         3       2   \n",
       "312                   5                     1         2       1   \n",
       "..                  ...                   ...       ...     ...   \n",
       "57                    2                  None         3       0   \n",
       "56                    3                  None         3       0   \n",
       "509                   2                  None         3       0   \n",
       "342                   3                  None         3       0   \n",
       "45                    1                  None         3       0   \n",
       "\n",
       "     target_IS_count  reference_IS_count  sig_ratio target reference  chrom  \n",
       "430             2518                   4   1.000000    CAR     NoCAR   chr4  \n",
       "360                1                1254   0.500000  NoCAR       CAR   chr9  \n",
       "556                4                2518   1.000000  NoCAR       CAR   chr4  \n",
       "260             1597                   5   0.666667  NoCAR       CAR   chr6  \n",
       "312             1254                   1   0.500000    CAR     NoCAR   chr9  \n",
       "..               ...                 ...        ...    ...       ...    ...  \n",
       "57                 3                   0   0.000000  NoCAR       CAR  chr14  \n",
       "56                 3                   0   0.000000  NoCAR       CAR  chr14  \n",
       "509                3                   0   0.000000  NoCAR       CAR   chr4  \n",
       "342                3                   0   0.000000  NoCAR       CAR   chr9  \n",
       "45                 3                   0   0.000000  NoCAR       CAR  chr16  \n",
       "\n",
       "[570 rows x 19 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "display(sig_df.sort_values(\"binomial\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the saved data\n",
    "# output_res = Path(\"/research/labs/immunology/rogerslm/m277102/projects/2023_SB/output/GRCm39/results-analysis/CAR-NoCAR\")  # mforge\n",
    "# output_res = Path(\"/research/labs/immunology/rogerslm/m277102/projects/2023_SB/output/GRCm39/results-analysis/ACF-SCF\")  # mforge\n",
    "# output_res = Path(\"/project/cs-myers/MathewF/projects/Laura-SB-Analysis/2023-SB-screen/output/GRCm39/results-analysis/CAR-NoCAR\")  # cse\n",
    "# output_res = Path(\"/project/cs-myers/MathewF/projects/Laura-SB-Analysis/2023-SB-screen/output/GRCm39/results-analysis/ACF-SCF\")  # cse\n",
    "\n",
    "\n",
    "TA_df = pd.read_csv(output_res / \"TA.tsv\", sep=\"\\t\")\n",
    "overall_df = pd.read_csv(output_res / \"overall.tsv\", sep=\"\\t\")\n",
    "sig_df = pd.read_csv(output_res / \"sig.tsv\", sep=\"\\t\")\n",
    "genomic_features_df = pd.read_csv(output_res / \"genomic_features.tsv\", sep=\"\\t\")\n",
    "graph_stats_df = pd.read_csv(output_res / \"graph_stats.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_index</th>\n",
       "      <th>reference_index</th>\n",
       "      <th>target_pos_min</th>\n",
       "      <th>target_pos_max</th>\n",
       "      <th>reference_pos_min</th>\n",
       "      <th>reference_pos_max</th>\n",
       "      <th>mannwhitneyu</th>\n",
       "      <th>ranksums</th>\n",
       "      <th>wilcoxon</th>\n",
       "      <th>binomial</th>\n",
       "      <th>total_IS</th>\n",
       "      <th>sig_IS</th>\n",
       "      <th>target_IS_count</th>\n",
       "      <th>reference_IS_count</th>\n",
       "      <th>sig_ratio</th>\n",
       "      <th>target</th>\n",
       "      <th>reference</th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40034220</td>\n",
       "      <td>40034221</td>\n",
       "      <td>40034220.0</td>\n",
       "      <td>40034220.0</td>\n",
       "      <td>0.414216</td>\n",
       "      <td>0.245278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>189</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr19</td>\n",
       "      <td>Cyp2c37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9362961</td>\n",
       "      <td>9362961</td>\n",
       "      <td>9362961.0</td>\n",
       "      <td>9362961.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.011719</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr19</td>\n",
       "      <td>Cyp2c37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40034220</td>\n",
       "      <td>40034221</td>\n",
       "      <td>40034220.0</td>\n",
       "      <td>40034220.0</td>\n",
       "      <td>0.414216</td>\n",
       "      <td>0.245278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>189</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr19</td>\n",
       "      <td>Cyp2c50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9362961</td>\n",
       "      <td>9362961</td>\n",
       "      <td>9362961.0</td>\n",
       "      <td>9362961.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.011719</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr19</td>\n",
       "      <td>Cyp2c50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40034220</td>\n",
       "      <td>40034221</td>\n",
       "      <td>40034220.0</td>\n",
       "      <td>40034220.0</td>\n",
       "      <td>0.414216</td>\n",
       "      <td>0.245278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>189</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr19</td>\n",
       "      <td>Cyp2c54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9362961</td>\n",
       "      <td>9362961</td>\n",
       "      <td>9362961.0</td>\n",
       "      <td>9362961.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.011719</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr19</td>\n",
       "      <td>Cyp2c54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>81</td>\n",
       "      <td>22.0</td>\n",
       "      <td>103793432</td>\n",
       "      <td>103793432</td>\n",
       "      <td>103793432.0</td>\n",
       "      <td>103816367.0</td>\n",
       "      <td>0.184039</td>\n",
       "      <td>0.12663</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>504</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "      <td>Dab1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>152</td>\n",
       "      <td>127.0</td>\n",
       "      <td>104512600</td>\n",
       "      <td>104512600</td>\n",
       "      <td>104512600.0</td>\n",
       "      <td>104512600.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>49</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CAR</td>\n",
       "      <td>NoCAR</td>\n",
       "      <td>chr4</td>\n",
       "      <td>Dab1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   target_index reference_index target_pos_min target_pos_max  \\\n",
       "76            2             4.0       40034220       40034221   \n",
       "77            4             2.0        9362961        9362961   \n",
       "78            2             4.0       40034220       40034221   \n",
       "79            4             2.0        9362961        9362961   \n",
       "80            2             4.0       40034220       40034221   \n",
       "81            4             2.0        9362961        9362961   \n",
       "83           81            22.0      103793432      103793432   \n",
       "84          152           127.0      104512600      104512600   \n",
       "\n",
       "   reference_pos_min reference_pos_max mannwhitneyu  ranksums wilcoxon  \\\n",
       "76        40034220.0        40034220.0     0.414216  0.245278      NaN   \n",
       "77         9362961.0         9362961.0          1.0  0.317311      NaN   \n",
       "78        40034220.0        40034220.0     0.414216  0.245278      NaN   \n",
       "79         9362961.0         9362961.0          1.0  0.317311      NaN   \n",
       "80        40034220.0        40034220.0     0.414216  0.245278      NaN   \n",
       "81         9362961.0         9362961.0          1.0  0.317311      NaN   \n",
       "83       103793432.0       103816367.0     0.184039   0.12663      NaN   \n",
       "84       104512600.0       104512600.0          1.0  0.317311      NaN   \n",
       "\n",
       "    binomial total_IS sig_IS target_IS_count reference_IS_count sig_ratio  \\\n",
       "76       0.0        2      1             189                  1       0.5   \n",
       "77  0.011719        1      1              10                  1       1.0   \n",
       "78       0.0        2      1             189                  1       0.5   \n",
       "79  0.011719        1      1              10                  1       1.0   \n",
       "80       0.0        2      1             189                  1       0.5   \n",
       "81  0.011719        1      1              10                  1       1.0   \n",
       "83       0.0        3      2               3                504  0.666667   \n",
       "84       0.0        1      1              49                  3       1.0   \n",
       "\n",
       "   target reference  chrom     gene  \n",
       "76    CAR     NoCAR  chr19  Cyp2c37  \n",
       "77    CAR     NoCAR  chr19  Cyp2c37  \n",
       "78    CAR     NoCAR  chr19  Cyp2c50  \n",
       "79    CAR     NoCAR  chr19  Cyp2c50  \n",
       "80    CAR     NoCAR  chr19  Cyp2c54  \n",
       "81    CAR     NoCAR  chr19  Cyp2c54  \n",
       "83    CAR     NoCAR   chr4     Dab1  \n",
       "84    CAR     NoCAR   chr4     Dab1  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assert len(candidate_genes) == len(output_genes)\n",
    "# TODO: 10/17/23\n",
    "# this assertion should not happen (or does it...?). Somehow there are genes that have gotten through the duplicate removal process\n",
    "arr, counts = np.unique(candidate_genes[\"gene\"].to_numpy(), return_counts=True)\n",
    "candidate_genes[candidate_genes[\"gene\"].isin(arr[counts == 2])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# validate gene candidates in a gene-centric way\n",
    "\n",
    "test for significance in different ways\n",
    "1. pCIS-case range\n",
    "2. pCIS-control range\n",
    "3. union of pCIS's\n",
    "4. intersection of pCIS's\n",
    "5. gene range (if able)\n",
    "6. gene range +50kb from both ends (if able)\n",
    "\n",
    "What statistical test should be used? For comparing the distance of two probability distributons, maybe look into Kullback–Leibler divergence?\n",
    "\n",
    "evaluate overall significance at different levels/amounts. What does the overall list look like if all tests must be significant?\n",
    "\n",
    "save this new conservative list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_index</th>\n",
       "      <th>reference_index</th>\n",
       "      <th>target_pos_min</th>\n",
       "      <th>target_pos_max</th>\n",
       "      <th>reference_pos_min</th>\n",
       "      <th>reference_pos_max</th>\n",
       "      <th>mannwhitneyu</th>\n",
       "      <th>ranksums</th>\n",
       "      <th>wilcoxon</th>\n",
       "      <th>binomial</th>\n",
       "      <th>...</th>\n",
       "      <th>union-binomial</th>\n",
       "      <th>intersect-mannwhitneyu</th>\n",
       "      <th>intersect-ranksums</th>\n",
       "      <th>intersect-binomial</th>\n",
       "      <th>gene-mannwhitneyu</th>\n",
       "      <th>gene-ranksums</th>\n",
       "      <th>gene-binomial</th>\n",
       "      <th>gene_extended-mannwhitneyu</th>\n",
       "      <th>gene_extended-ranksums</th>\n",
       "      <th>gene_extended-binomial</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>103846357</td>\n",
       "      <td>103846381</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.021071</td>\n",
       "      <td>0.020921</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.021071</td>\n",
       "      <td>0.020921</td>\n",
       "      <td>2.672765e-51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>138825098</td>\n",
       "      <td>138825098</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>3.125000e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>54</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27307799</td>\n",
       "      <td>27307799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>1.220703e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50676643</td>\n",
       "      <td>50676646</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.059346</td>\n",
       "      <td>0.049535</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.059346</td>\n",
       "      <td>0.049535</td>\n",
       "      <td>1.033976e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>145654994</td>\n",
       "      <td>145654994</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>1.525879e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147755841</td>\n",
       "      <td>147755844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.220671</td>\n",
       "      <td>0.121335</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.063603</td>\n",
       "      <td>0.049535</td>\n",
       "      <td>6.103516e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>174</td>\n",
       "      <td>NaN</td>\n",
       "      <td>145394123</td>\n",
       "      <td>145394123</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.007812</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>7.812500e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>146675191</td>\n",
       "      <td>146675191</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000977</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.317311</td>\n",
       "      <td>9.765625e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>126941561</td>\n",
       "      <td>126941597</td>\n",
       "      <td>126941571.0</td>\n",
       "      <td>126941601.0</td>\n",
       "      <td>0.325364</td>\n",
       "      <td>0.306685</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.513028e-12</td>\n",
       "      <td>0.883853</td>\n",
       "      <td>0.77283</td>\n",
       "      <td>0.837556</td>\n",
       "      <td>0.325364</td>\n",
       "      <td>0.306685</td>\n",
       "      <td>5.513028e-12</td>\n",
       "      <td>0.325364</td>\n",
       "      <td>0.306685</td>\n",
       "      <td>5.513028e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>126941561</td>\n",
       "      <td>126941597</td>\n",
       "      <td>126941571.0</td>\n",
       "      <td>126941601.0</td>\n",
       "      <td>0.325364</td>\n",
       "      <td>0.306685</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.513028e-12</td>\n",
       "      <td>0.883853</td>\n",
       "      <td>0.77283</td>\n",
       "      <td>0.837556</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.325364</td>\n",
       "      <td>0.306685</td>\n",
       "      <td>5.513028e-12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>368 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    target_index reference_index target_pos_min target_pos_max  \\\n",
       "0              2             NaN      103846357      103846381   \n",
       "1             58             NaN      138825098      138825098   \n",
       "2             54             NaN       27307799       27307799   \n",
       "3              6             NaN       50676643       50676646   \n",
       "4             10             NaN      145654994      145654994   \n",
       "..           ...             ...            ...            ...   \n",
       "363           29             NaN      147755841      147755844   \n",
       "364          174             NaN      145394123      145394123   \n",
       "365          196             NaN      146675191      146675191   \n",
       "366            8             0.0      126941561      126941597   \n",
       "367            8             0.0      126941561      126941597   \n",
       "\n",
       "    reference_pos_min reference_pos_max mannwhitneyu  ranksums wilcoxon  \\\n",
       "0                 NaN               NaN     0.021071  0.020921      NaN   \n",
       "1                 NaN               NaN          1.0  0.317311      NaN   \n",
       "2                 NaN               NaN          1.0  0.317311      NaN   \n",
       "3                 NaN               NaN     0.059346  0.049535      NaN   \n",
       "4                 NaN               NaN          1.0  0.317311      NaN   \n",
       "..                ...               ...          ...       ...      ...   \n",
       "363               NaN               NaN     0.220671  0.121335      NaN   \n",
       "364               NaN               NaN          1.0  0.317311      NaN   \n",
       "365               NaN               NaN          1.0  0.317311      NaN   \n",
       "366       126941571.0       126941601.0     0.325364  0.306685      NaN   \n",
       "367       126941571.0       126941601.0     0.325364  0.306685      NaN   \n",
       "\n",
       "     binomial  ... union-binomial intersect-mannwhitneyu intersect-ranksums  \\\n",
       "0         0.0  ...            NaN                    NaN                NaN   \n",
       "1     0.03125  ...            NaN                    NaN                NaN   \n",
       "2    0.000122  ...            NaN                    NaN                NaN   \n",
       "3         0.0  ...            NaN                    NaN                NaN   \n",
       "4    0.000015  ...            NaN                    NaN                NaN   \n",
       "..        ...  ...            ...                    ...                ...   \n",
       "363  0.000244  ...            NaN                    NaN                NaN   \n",
       "364  0.007812  ...            NaN                    NaN                NaN   \n",
       "365  0.000977  ...            NaN                    NaN                NaN   \n",
       "366       0.0  ...   5.513028e-12               0.883853            0.77283   \n",
       "367       0.0  ...   5.513028e-12               0.883853            0.77283   \n",
       "\n",
       "    intersect-binomial gene-mannwhitneyu gene-ranksums gene-binomial  \\\n",
       "0                  NaN               NaN           NaN           NaN   \n",
       "1                  NaN               NaN           NaN           NaN   \n",
       "2                  NaN               NaN           NaN           NaN   \n",
       "3                  NaN               NaN           NaN           NaN   \n",
       "4                  NaN               NaN           NaN           NaN   \n",
       "..                 ...               ...           ...           ...   \n",
       "363                NaN               NaN           NaN           NaN   \n",
       "364                NaN               NaN           NaN           NaN   \n",
       "365                NaN               NaN           NaN           NaN   \n",
       "366           0.837556          0.325364      0.306685  5.513028e-12   \n",
       "367           0.837556               NaN           NaN           NaN   \n",
       "\n",
       "    gene_extended-mannwhitneyu gene_extended-ranksums  gene_extended-binomial  \n",
       "0                     0.021071               0.020921            2.672765e-51  \n",
       "1                     1.000000               0.317311            3.125000e-02  \n",
       "2                     1.000000               0.317311            1.220703e-04  \n",
       "3                     0.059346               0.049535            1.033976e-25  \n",
       "4                     1.000000               0.317311            1.525879e-05  \n",
       "..                         ...                    ...                     ...  \n",
       "363                   0.063603               0.049535            6.103516e-05  \n",
       "364                   1.000000               0.317311            7.812500e-03  \n",
       "365                   1.000000               0.317311            9.765625e-04  \n",
       "366                   0.325364               0.306685            5.513028e-12  \n",
       "367                   0.325364               0.306685            5.513028e-12  \n",
       "\n",
       "[368 rows x 37 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def get_pCIS_insertions(row, pcIS_type, ta_df, range_min=None, range_max=None): \n",
    "    # pcIS_type can be target or reference\n",
    "    tmp1 = ta_df[ta_df[\"target\"] == row[\"target\"]]\n",
    "    tmp2 = tmp1[tmp1[f\"target_index\"] == row[f\"target_index\"]]\n",
    "    tmp3 = tmp2[tmp2[\"chrom\"] == row[\"chrom\"]]\n",
    "    \n",
    "    if (range_min == None) and (range_max == None):\n",
    "        TAs = tmp3[(tmp3[\"pos\"] >= row[f\"{pcIS_type}_pos_min\"]) & (tmp3[\"pos\"] <= row[f\"{pcIS_type}_pos_max\"])]\n",
    "        return TAs\n",
    "    elif (range_min != None) and (range_max != None):\n",
    "        TAs = tmp3[(tmp3[\"pos\"] >= range_min) & (tmp3[\"pos\"] <= range_max)]\n",
    "        return TAs\n",
    "    else:\n",
    "        print(f\"Nope, can't do that in get_pCIS_insertions: range_min={range_min}, range_max={range_max}\")\n",
    "\n",
    "def get_pCIS_stats(tmp, prepend):\n",
    "    if not len(tmp):\n",
    "        mwu, rs, wc, binom = np.NAN, np.NAN, np.NAN, np.NAN\n",
    "    else:\n",
    "        mwu = mannwhitneyu(tmp[\"target_count\"], tmp[\"reference_count\"]).pvalue \n",
    "        rs = ranksums(tmp[\"target_count\"], tmp[\"reference_count\"]).pvalue\n",
    "        # wc = wilcoxon(tmp[\"target_count\"], tmp[\"reference_count\"]).pvalue if (tmp[\"target_count\"] - tmp[\"reference_count\"]).sum() != 0 and len(tmp) >= 10 else np.nan \n",
    "        # binom = binomtest(tmp[\"target_count\"].sum(), tmp[\"target_count\"].sum() + tmp[\"reference_count\"].sum(), 0.5).pvalue\n",
    "        \n",
    "    out_df = pd.DataFrame({\n",
    "        f\"{prepend}-mannwhitneyu\": [mwu],  # https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mannwhitneyu.html#scipy-stats-mannwhitneyu\n",
    "        f\"{prepend}-ranksums\": [rs],       # https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ranksums.html\n",
    "        # f\"{prepend}-wilcoxon\": [wc],       # https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wilcoxon.html#scipy.stats.wilcoxon\n",
    "        # f\"{prepend}-binomial\": [binom],\n",
    "        })\n",
    "    \n",
    "    return out_df\n",
    "\n",
    "\n",
    "# for each gene (or pCIS???)\n",
    "new_gene_candidates = []\n",
    "for i, row in candidate_genes.iterrows():\n",
    "    # if pd.isna(row[\"reference_index\"]):\n",
    "    #     continue\n",
    "    # if i != 21:\n",
    "    #     continue\n",
    "    \n",
    "    ### test for significance in the 6 ways using the 4 stat methods\n",
    "    \n",
    "    # target pCIS range\n",
    "    target_TAs = get_pCIS_insertions(row, \"target\", TA_df)\n",
    "    target_out = get_pCIS_stats(target_TAs, row[\"target\"])\n",
    "\n",
    "    \n",
    "    # check ref range, union, and intersection if there is a ref index\n",
    "    if pd.notna(row[\"reference_index\"]):\n",
    "        \n",
    "        # reference pCIS range if possible\n",
    "        ref_TAs = get_pCIS_insertions(row, \"reference\", TA_df)\n",
    "        ref_out = get_pCIS_stats(ref_TAs, row[\"reference\"])\n",
    "\n",
    "        # union of pCIS range if possible\n",
    "        union_TAs = get_pCIS_insertions(row, \"target\", TA_df,\n",
    "                                        min(row[\"target_pos_min\"], row[\"reference_pos_min\"]),\n",
    "                                        max(row[\"target_pos_max\"], row[\"reference_pos_max\"]),\n",
    "                                        )\n",
    "        union_out = get_pCIS_stats(union_TAs, \"union\")\n",
    "        \n",
    "        # intersection of pCIS range if possible\n",
    "        intersect_TAs = get_pCIS_insertions(row, \"target\", TA_df,\n",
    "                                        max(row[\"target_pos_min\"], row[\"reference_pos_min\"]),\n",
    "                                        min(row[\"target_pos_max\"], row[\"reference_pos_max\"]),\n",
    "                                        )\n",
    "        intersect_out = get_pCIS_stats(intersect_TAs, \"intersect\")\n",
    "        \n",
    "    else:\n",
    "        ref_out = get_pCIS_stats([], row[\"reference\"])\n",
    "        union_out = get_pCIS_stats([], \"union\")\n",
    "        intersect_out = get_pCIS_stats([], \"intersect\")\n",
    "\n",
    "        \n",
    "    # gene range (TODO: check for this for all significant results, not just the ones near genes)\n",
    "    gene_row = annot_df[annot_df[\"Marker Symbol\"] == row[\"gene\"]]\n",
    "    gene_start = gene_row[\"genome coordinate start\"].tolist()[0]\n",
    "    gene_end = gene_row[\"genome coordinate end\"].tolist()[0]\n",
    "    gene_TAs = TA_df[(TA_df[\"chrom\"] == row[\"chrom\"]) & (TA_df[\"pos\"] >= gene_start) & (TA_df[\"pos\"] <= gene_end)]\n",
    "    gene_TAs = gene_TAs.drop_duplicates(subset=[\"pos\"], keep='first')\n",
    "    genes_out = get_pCIS_stats(gene_TAs, \"gene\")\n",
    "        \n",
    "    # gene range +/- 50 kb\n",
    "    gene_extended_TAs = TA_df[(TA_df[\"chrom\"] == row[\"chrom\"]) & (TA_df[\"pos\"] >= (gene_start - 50000)) & (TA_df[\"pos\"] <= (gene_end + 50000))]\n",
    "    gene_extended_TAs = gene_extended_TAs.drop_duplicates(subset=[\"pos\"], keep='first')\n",
    "    genes_extended_out = get_pCIS_stats(gene_extended_TAs, \"gene_extended\")\n",
    "\n",
    "    # save as new list\n",
    "    row_out = pd.concat([target_out, ref_out, union_out, intersect_out, genes_out, genes_extended_out], axis=1)\n",
    "    new_gene_candidates.append(pd.concat([row.to_frame().T.reset_index(drop=True), row_out], axis=1))\n",
    "    \n",
    "# view output\n",
    "new_gene_candidates_df = pd.concat(new_gene_candidates, ignore_index=True)\n",
    "display(new_gene_candidates_df)\n",
    "\n",
    "# save new list\n",
    "new_gene_candidates_df.to_csv(output_res / \"validated_candidate_genes.tsv\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What about breaking up large CISs? It appears they are too large and span multiple genes\n",
    "\n",
    "# Or do now we check each individual gene that was found to further refine our search?\n",
    "\n",
    "# What's stopping me from testing literally every gene with +/-50 kb region of insertions with a ranksums test?\n",
    "# because we don't want this to be gene centric, however, it's not a bad way to validate the candidate gene list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: check for each gene with an extended proomoter the case vs controls insertions\n",
    "# check with ranksums test as well as binomial test\n",
    "# What does the candidate gene list look like now?\n",
    "\n",
    "\n",
    "# TODO: I need to go through my code again. Go step by step for each function in another notebook\n",
    "# double check this code and cis_networks.py\n",
    "\n",
    "\n",
    "# TODO: rerun without mapq thresholding in preprocess_reads and preprocess_insertions\n",
    "\n",
    "\n",
    "# TODO: should we be using normalized read counts? idk, what stage would this occur at\n",
    "# the variation would come from read depth, so we would have to normalize based on this BEFORE preprocess_reads.py\n",
    "# What is done with RNAseq? How to normalize on read depth while still keeping count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# volcano plot of p-values and LFC for each CIS if wanted\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output list of pCIS (now can be called CIS) that are sig.\n",
    "# per CIS, output list of TAs that are sig.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# take identified candidate CIS/genes and check for co-occurrence with all other CIS/genes\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "netcis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
